diff --git a/Tests/test_image.py b/Tests/test_image.py
index c2e850c36..5f8e94884 100644
--- a/Tests/test_image.py
+++ b/Tests/test_image.py
@@ -182,6 +182,22 @@ class TestImage:
         with Image.open(out) as reloaded:
             assert_image_equal(im, reloaded)
 
+    def test_dry_run(self, tmp_path: Path) -> None:
+        out = tmp_path / "temp.png"
+        im = hopper()
+        im.save(out, dry_run=True)
+        assert not out.exists()
+
+        im.save(out)
+        assert out.exists()
+
+    def test_dry_run_raises_exception(self, tmp_path: Path) -> None:
+        out = tmp_path / "temp.jpg"
+        im = hopper("RGBA")
+        with pytest.raises(OSError):
+            im.save(out, dry_run=True)
+        assert not out.exists()
+
     def test_pathlib(self, tmp_path: Path) -> None:
         with Image.open(Path("Tests/images/multipage-mmap.tiff")) as im:
             assert im.mode == "P"
diff --git a/src/PIL/Image.py b/src/PIL/Image.py
index 233df592c..07823a7fe 100644
--- a/src/PIL/Image.py
+++ b/src/PIL/Image.py
@@ -2467,7 +2467,12 @@ class Image:
         )
 
     def save(
-        self, fp: StrOrBytesPath | IO[bytes], format: str | None = None, **params: Any
+        self,
+        fp: StrOrBytesPath | IO[bytes],
+        format: str | None = None,
+        *,
+        dry_run: bool = False,
+        **params: Any,
     ) -> None:
         """
         Saves this image under the given filename.  If no format is
@@ -2490,6 +2495,10 @@ class Image:
            format to use is determined from the filename extension.
            If a file object was used instead of a filename, this
            parameter should always be used.
+        :param dry_run: If true, the image is processed as if it were being
+           saved, including any format conversion, validation and errors that
+           this would raise, but nothing is written to disk and ``fp`` is
+           otherwise unused.
         :param params: Extra parameters to the image writer. These can also be
            set on the image itself through ``encoderinfo``. This is useful when
            saving multiple images::
@@ -2559,7 +2568,9 @@ class Image:
             save_handler = SAVE[format.upper()]
 
         created = False
-        if open_fp:
+        if dry_run:
+            fp = io.BytesIO()
+        elif open_fp:
             created = not os.path.exists(filename)
             if params.get("append", False):
                 # Open also for reading ("+"), because TIFF save_all
@@ -2573,7 +2584,7 @@ class Image:
         try:
             save_handler(self, fp, filename)
         except Exception:
-            if open_fp:
+            if open_fp and not dry_run:
                 fp.close()
             if created:
                 try:
@@ -2586,7 +2597,7 @@ class Image:
                 del self.encoderinfo
             except AttributeError:
                 pass
-        if open_fp:
+        if open_fp and not dry_run:
             fp.close()
 
     def seek(self, frame: int) -> None:
