diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..a6d410a 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,4 +1,5 @@
 import json
+import re
 from typing import Any, Callable, TypeVar, overload
 from uuid import UUID
 
@@ -145,3 +146,50 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+HashTypes = Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, either in general or of a particular algorithm.
+    """
+
+    hash_lengths = {
+        'md5': 32,
+        'sha1': 40,
+        'sha224': 56,
+        'sha256': 64,
+        'sha384': 96,
+        'sha512': 128,
+    }
+
+    def __init__(self, algorithm: HashTypes):
+        """
+        Args:
+            algorithm: The algorithm to use for the hash, must be one of `'md5'`, `'sha1'`, `'sha224'`,
+                `'sha256'`, `'sha384'`, or `'sha512'`.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '5eb63bbbe01eeed093cb22bb8f5acdc3' == IsHash('md5')
+        assert '2aae6c35c94fcfb415dbe95f408b9ce91ee846ed' == IsHash('sha1')
+        assert 'not-a-hash' != IsHash('md5')
+        ```
+        """
+        self.algorithm = algorithm.lower()
+        if self.algorithm not in self.hash_lengths:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        self.expected_length = self.hash_lengths[self.algorithm]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        if len(other) != self.expected_length:
+            return False
+
+        return bool(re.fullmatch('[0-9a-fA-F]+', other))
