diff --git a/src/jinja2/ext.py b/src/jinja2/ext.py
index d21b83aa..72ac5837 100644
--- a/src/jinja2/ext.py
+++ b/src/jinja2/ext.py
@@ -200,6 +200,42 @@ def _make_new_ngettext(func: t.Callable[[str, str, int], str]) -> t.Callable[...
     return ngettext
 
 
+def _make_new_dgettext(func: t.Callable[[str, str], str]) -> t.Callable[..., str]:
+    @pass_context
+    def dgettext(
+        __context: Context, __domain: str, __string: str, **variables: t.Any
+    ) -> str:
+        rv = __context.call(func, __domain, __string)
+        if __context.eval_ctx.autoescape:
+            rv = Markup(rv)
+        # Always treat as a format string, see gettext comment above.
+        return rv % variables  # type: ignore
+
+    return dgettext
+
+
+def _make_new_dngettext(
+    func: t.Callable[[str, str, str, int], str]
+) -> t.Callable[..., str]:
+    @pass_context
+    def dngettext(
+        __context: Context,
+        __domain: str,
+        __singular: str,
+        __plural: str,
+        __num: int,
+        **variables: t.Any,
+    ) -> str:
+        variables.setdefault("num", __num)
+        rv = __context.call(func, __domain, __singular, __plural, __num)
+        if __context.eval_ctx.autoescape:
+            rv = Markup(rv)
+        # Always treat as a format string, see gettext comment above.
+        return rv % variables  # type: ignore
+
+    return dngettext
+
+
 def _make_new_pgettext(func: t.Callable[[str, str], str]) -> t.Callable[..., str]:
     @pass_context
     def pgettext(
@@ -280,8 +316,16 @@ class InternationalizationExtension(Extension):
 
         pgettext = getattr(translations, "pgettext", None)
         npgettext = getattr(translations, "npgettext", None)
+        dgettext = getattr(translations, "dgettext", None)
+        dngettext = getattr(translations, "dngettext", None)
         self._install_callables(
-            gettext, ngettext, newstyle=newstyle, pgettext=pgettext, npgettext=npgettext
+            gettext,
+            ngettext,
+            newstyle=newstyle,
+            pgettext=pgettext,
+            npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _install_null(self, newstyle: t.Optional[bool] = None) -> None:
@@ -304,12 +348,17 @@ class InternationalizationExtension(Extension):
             def npgettext(c: str, s: str, p: str, n: int) -> str:
                 return s if n == 1 else p
 
+        dgettext = getattr(translations, "dgettext", None)
+        dngettext = getattr(translations, "dngettext", None)
+
         self._install_callables(
             gettext=translations.gettext,
             ngettext=translations.ngettext,
             newstyle=newstyle,
             pgettext=pgettext,
             npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _install_callables(
@@ -319,12 +368,29 @@ class InternationalizationExtension(Extension):
         newstyle: t.Optional[bool] = None,
         pgettext: t.Optional[t.Callable[[str, str], str]] = None,
         npgettext: t.Optional[t.Callable[[str, str, str, int], str]] = None,
+        dgettext: t.Optional[t.Callable[[str, str], str]] = None,
+        dngettext: t.Optional[t.Callable[[str, str, str, int], str]] = None,
     ) -> None:
         if newstyle is not None:
             self.environment.newstyle_gettext = newstyle  # type: ignore
+
+        if dgettext is None:
+            _base_gettext = gettext
+
+            def dgettext(domain: str, message: str) -> str:
+                return _base_gettext(message)
+
+        if dngettext is None:
+            _base_ngettext = ngettext
+
+            def dngettext(domain: str, singular: str, plural: str, num: int) -> str:
+                return _base_ngettext(singular, plural, num)
+
         if self.environment.newstyle_gettext:  # type: ignore
             gettext = _make_new_gettext(gettext)
             ngettext = _make_new_ngettext(ngettext)
+            dgettext = _make_new_dgettext(dgettext)
+            dngettext = _make_new_dngettext(dngettext)
 
             if pgettext is not None:
                 pgettext = _make_new_pgettext(pgettext)
@@ -333,11 +399,23 @@ class InternationalizationExtension(Extension):
                 npgettext = _make_new_npgettext(npgettext)
 
         self.environment.globals.update(
-            gettext=gettext, ngettext=ngettext, pgettext=pgettext, npgettext=npgettext
+            gettext=gettext,
+            ngettext=ngettext,
+            pgettext=pgettext,
+            npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _uninstall(self, translations: "_SupportedTranslations") -> None:
-        for key in ("gettext", "ngettext", "pgettext", "npgettext"):
+        for key in (
+            "gettext",
+            "ngettext",
+            "pgettext",
+            "npgettext",
+            "dgettext",
+            "dngettext",
+        ):
             self.environment.globals.pop(key, None)
 
     def _extract(
@@ -363,6 +441,7 @@ class InternationalizationExtension(Extension):
         plural_expr_assignment: t.Optional[nodes.Assign] = None
         variables: t.Dict[str, nodes.Expr] = {}
         trimmed = None
+        domain: t.Optional[str] = None
         while parser.stream.current.type != "block_end":
             if variables:
                 parser.stream.expect("comma")
@@ -380,7 +459,27 @@ class InternationalizationExtension(Extension):
                 )
 
             # expressions
-            if parser.stream.current.type == "assign":
+            if token.value == "domain" and parser.stream.current.type == "assign":
+                if domain is not None:
+                    parser.fail(
+                        "translation domain specified twice.",
+                        token.lineno,
+                        exc=TemplateAssertionError,
+                    )
+                next(parser.stream)
+                domain_token = parser.stream.current
+                domain_expr = parser.parse_expression()
+                if not isinstance(domain_expr, nodes.Const) or not isinstance(
+                    domain_expr.value, str
+                ):
+                    parser.fail(
+                        "translation domain must be a string literal constant.",
+                        domain_token.lineno,
+                        exc=TemplateAssertionError,
+                    )
+                domain = domain_expr.value
+                continue
+            elif parser.stream.current.type == "assign":
                 next(parser.stream)
                 variables[token.value] = var = parser.parse_expression()
             elif trimmed is None and token.value in ("trimmed", "notrimmed"):
@@ -459,6 +558,7 @@ class InternationalizationExtension(Extension):
             plural_expr,
             bool(referenced),
             num_called_num and have_plural,
+            domain,
         )
         node.set_lineno(lineno)
         if plural_expr_assignment is not None:
@@ -514,6 +614,7 @@ class InternationalizationExtension(Extension):
         plural_expr: t.Optional[nodes.Expr],
         vars_referenced: bool,
         num_called_num: bool,
+        domain: t.Optional[str] = None,
     ) -> nodes.Output:
         """Generates a useful node from the data provided."""
         newstyle = self.environment.newstyle_gettext  # type: ignore
@@ -526,17 +627,23 @@ class InternationalizationExtension(Extension):
             if plural:
                 plural = plural.replace("%%", "%")
 
+        domain_args = [nodes.Const(domain)] if domain is not None else []
+
         # singular only:
         if plural_expr is None:
-            gettext = nodes.Name("gettext", "load")
-            node = nodes.Call(gettext, [nodes.Const(singular)], [], None, None)
+            func_name = "dgettext" if domain is not None else "gettext"
+            gettext = nodes.Name(func_name, "load")
+            node = nodes.Call(
+                gettext, domain_args + [nodes.Const(singular)], [], None, None
+            )
 
         # singular and plural
         else:
-            ngettext = nodes.Name("ngettext", "load")
+            func_name = "dngettext" if domain is not None else "ngettext"
+            ngettext = nodes.Name(func_name, "load")
             node = nodes.Call(
                 ngettext,
-                [nodes.Const(singular), nodes.Const(plural), plural_expr],
+                domain_args + [nodes.Const(singular), nodes.Const(plural), plural_expr],
                 [],
                 None,
                 None,
