diff --git a/Tests/test_image.py b/Tests/test_image.py
index c2e850c36..464cb2350 100644
--- a/Tests/test_image.py
+++ b/Tests/test_image.py
@@ -258,6 +258,22 @@ class TestImage:
             assert im.readonly
             im.save(temp_file)
 
+    def test_readonly_save_frombuffer(self, tmp_path: Path) -> None:
+        temp_file = tmp_path / "temp.bmp"
+
+        data = bytearray(b"\xff")
+        with Image.frombuffer("L", (1, 1), data) as im:
+            assert im.readonly
+
+            data[0] = 128
+            assert im.getdata()[0] == 128
+
+            im.save(temp_file)
+            assert im.readonly
+
+            data[0] = 64
+            assert im.getdata()[0] == 64
+
     def test_dump(self, tmp_path: Path) -> None:
         im = Image.new("L", (10, 10))
         im._dump(str(tmp_path / "temp_L.ppm"))
diff --git a/src/PIL/Image.py b/src/PIL/Image.py
index 233df592c..41a806404 100644
--- a/src/PIL/Image.py
+++ b/src/PIL/Image.py
@@ -2540,8 +2540,11 @@ class Image:
                 msg = f"unknown file extension: {ext}"
                 raise ValueError(msg) from e
 
-        # may mutate self!
-        self._ensure_mutable()
+        if filename and getattr(self, "filename", None) == filename:
+            # may mutate self!
+            self._ensure_mutable()
+        else:
+            self.load()
 
         save_all = params.pop("save_all", None)
         self.encoderinfo = {**getattr(self, "encoderinfo", {}), **params}
