diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..2bfb95a 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,4 +1,5 @@
 import json
+import re
 from typing import Any, Callable, TypeVar, overload
 from uuid import UUID
 
@@ -145,3 +146,50 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+HashTypes = Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+
+_hash_lengths: 'dict[str, int]' = {
+    'md5': 32,
+    'sha1': 40,
+    'sha224': 56,
+    'sha256': 64,
+    'sha384': 96,
+    'sha512': 128,
+}
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash of a given algorithm/hash type.
+    """
+
+    def __init__(self, algorithm: HashTypes):
+        """
+        Args:
+            algorithm: The hash algorithm to check, must be one of `'md5'`, `'sha1'`, `'sha224'`, `'sha256'`,
+                `'sha384'` or `'sha512'`.
+
+        ```py title="IsHash"
+        import hashlib
+        from dirty_equals import IsHash
+
+        assert hashlib.md5(b'hello').hexdigest() == IsHash('md5')
+        assert hashlib.sha1(b'hello').hexdigest() == IsHash('sha1')
+        assert hashlib.sha256(b'hello').hexdigest() == IsHash('sha256')
+        assert 'not a hash' != IsHash('md5')
+        ```
+        """
+        self.algorithm = algorithm
+        self._length = _hash_lengths[algorithm.lower()]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        if len(other) != self._length:
+            return False
+
+        return bool(re.fullmatch('[0-9a-fA-F]+', other))
