diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..291e95c 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsPhoneNumber, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -67,7 +67,9 @@ __all__ = (
     'IsInstance',
     # other
     'FunctionCheck',
+    'IsHash',
     'IsJson',
+    'IsPhoneNumber',
     'IsUUID',
     # strings
     'IsStr',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..f48a29c 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -145,3 +145,51 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, for a given hash type.
+    """
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm to check the hash against, one of `'md5'`, `'sha1'`, `'sha224'`,
+                `'sha256'`, `'sha384'` or `'sha512'`.
+
+        ```py title="IsHash"
+        import hashlib
+        from dirty_equals import IsHash
+
+        assert hashlib.md5(b'hello').hexdigest() == IsHash('md5')
+        assert hashlib.sha256(b'hello').hexdigest() == IsHash('sha256')
+        assert 'not_a_hash' != IsHash('md5')
+        ```
+        """
+        self.algorithm = algorithm
+        self.algorithm_length = {
+            'md5': 32,
+            'sha1': 40,
+            'sha224': 56,
+            'sha256': 64,
+            'sha384': 96,
+            'sha512': 128,
+        }.get(algorithm.lower())
+        if self.algorithm_length is None:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        if len(other) != self.algorithm_length:
+            return False
+
+        try:
+            int(other, 16)
+        except ValueError:
+            return False
+
+        return True
