diff --git a/crates/typst/src/foundations/str.rs b/crates/typst/src/foundations/str.rs
index d90b6f206..1c18388bc 100644
--- a/crates/typst/src/foundations/str.rs
+++ b/crates/typst/src/foundations/str.rs
@@ -179,9 +179,18 @@ impl Str {
     /// Extracts the first grapheme cluster of the string.
     /// Fails with an error if the string is empty.
     #[func]
-    pub fn first(&self) -> StrResult<Str> {
-        self.0
-            .graphemes(true)
+    pub fn first(
+        &self,
+        /// A set of characters to strip from the start of the string before
+        /// extracting the first grapheme cluster.
+        #[named]
+        strip: Option<Str>,
+    ) -> StrResult<Str> {
+        let text = match &strip {
+            Some(strip) => self.0.trim_start_matches(|c| strip.as_str().contains(c)),
+            None => self.0.as_str(),
+        };
+        text.graphemes(true)
             .next()
             .map(Into::into)
             .ok_or_else(string_is_empty)
@@ -190,9 +199,18 @@ impl Str {
     /// Extracts the last grapheme cluster of the string.
     /// Fails with an error if the string is empty.
     #[func]
-    pub fn last(&self) -> StrResult<Str> {
-        self.0
-            .graphemes(true)
+    pub fn last(
+        &self,
+        /// A set of characters to strip from the end of the string before
+        /// extracting the last grapheme cluster.
+        #[named]
+        strip: Option<Str>,
+    ) -> StrResult<Str> {
+        let text = match &strip {
+            Some(strip) => self.0.trim_end_matches(|c| strip.as_str().contains(c)),
+            None => self.0.as_str(),
+        };
+        text.graphemes(true)
             .next_back()
             .map(Into::into)
             .ok_or_else(string_is_empty)
