diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..24b4f70 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -19,6 +19,7 @@ from ._numeric import (
     IsPositiveInt,
 )
 from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import IsHash
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +70,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..9183ea5 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -145,3 +145,48 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+HASH_ALGO_LENGTHS = {
+    'md5': 32,
+    'sha1': 40,
+    'sha224': 56,
+    'sha256': 64,
+    'sha384': 96,
+    'sha512': 128,
+}
+HEX_DIGITS = set('0123456789abcdefABCDEF')
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, for a given hashing algorithm.
+    """
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The hashing algorithm to check the hash against, must be one of
+                `'md5'`, `'sha1'`, `'sha224'`, `'sha256'`, `'sha384'` or `'sha512'`.
+
+        ```py title="IsHash"
+        import hashlib
+        from dirty_equals import IsHash
+
+        assert hashlib.md5(b'hello').hexdigest() == IsHash('md5')
+        assert hashlib.sha1(b'hello').hexdigest() == IsHash('sha1')
+        assert hashlib.sha256(b'hello').hexdigest() == IsHash('sha256')
+        assert 'not-a-hash' != IsHash('sha256')
+        ```
+        """
+        try:
+            self.hash_length = HASH_ALGO_LENGTHS[algorithm.lower()]
+        except KeyError:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}, must be one of {sorted(HASH_ALGO_LENGTHS)}')
+        self.algorithm = algorithm
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        return len(other) == self.hash_length and all(c in HEX_DIGITS for c in other)
