diff --git a/outlines/types/__init__.py b/outlines/types/__init__.py
index 85af9adc..61d1a107 100644
--- a/outlines/types/__init__.py
+++ b/outlines/types/__init__.py
@@ -68,3 +68,10 @@ email = Regex(
 isbn = Regex(
     r"(?:ISBN(?:-1[03])?:? )?(?=[0-9X]{10}$|(?=(?:[0-9]+[- ]){3})[- 0-9X]{13}$|97[89][0-9]{10}$|(?=(?:[0-9]+[- ]){4})[- 0-9]{17}$)(?:97[89][- ]?)?[0-9]{1,5}[- ]?[0-9]+[- ]?[0-9]+[- ]?[0-9X]"
 )
+
+# Matches a 16-digit credit card number, either with no separators, or with
+# groups of 4 digits separated consistently by spaces or by hyphens (but not
+# a mix of the two).
+credit_card = Regex(
+    r"(?:[0-9]{16}|[0-9]{4}(?: [0-9]{4}){3}|[0-9]{4}(?:-[0-9]{4}){3})"
+)
