diff --git a/llama-index-core/llama_index/core/base/llms/types.py b/llama-index-core/llama_index/core/base/llms/types.py
index dedf4ad1e..6036d22c8 100644
--- a/llama-index-core/llama_index/core/base/llms/types.py
+++ b/llama-index-core/llama_index/core/base/llms/types.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 import base64
 import filetype
+import os
 from binascii import Error as BinasciiError
 from enum import Enum
 from io import BytesIO
@@ -125,6 +126,19 @@ class ImageBlock(BaseModel):
             as_base64=as_base64,
         )
 
+    def resolve_image_with_size(self, as_base64: bool = False) -> tuple[BytesIO, int]:
+        """
+        Resolve an image such that PIL can read it, and also return its size in bytes.
+
+        Args:
+            as_base64 (bool): whether the resolved image should be returned as base64-encoded bytes
+
+        """
+        buffer = self.resolve_image(as_base64=as_base64)
+        size = buffer.seek(0, os.SEEK_END)
+        buffer.seek(0)
+        return buffer, size
+
 
 class AudioBlock(BaseModel):
     block_type: Literal["audio"] = "audio"
@@ -185,6 +199,19 @@ class AudioBlock(BaseModel):
             as_base64=as_base64,
         )
 
+    def resolve_audio_with_size(self, as_base64: bool = False) -> tuple[BytesIO, int]:
+        """
+        Resolve an audio such that PIL can read it, and also return its size in bytes.
+
+        Args:
+            as_base64 (bool): whether the resolved audio should be returned as base64-encoded bytes
+
+        """
+        buffer = self.resolve_audio(as_base64=as_base64)
+        size = buffer.seek(0, os.SEEK_END)
+        buffer.seek(0)
+        return buffer, size
+
 class DocumentBlock(BaseModel):
     block_type: Literal["document"] = "document"
     data: Optional[bytes] = None
@@ -222,6 +249,15 @@ class DocumentBlock(BaseModel):
             as_base64=False,
         )
 
+    def resolve_document_with_size(self) -> tuple[BytesIO, int]:
+        """
+        Resolve a document such that it is represented by a BufferIO object, and also return its size in bytes.
+        """
+        buffer = self.resolve_document()
+        size = buffer.seek(0, os.SEEK_END)
+        buffer.seek(0)
+        return buffer, size
+
     def _get_b64_string(self, data_buffer: BytesIO) -> str:
         """
         Get base64-encoded string from a BytesIO buffer.
