diff --git a/Tests/test_image.py b/Tests/test_image.py
index c2e850c36..50321c285 100644
--- a/Tests/test_image.py
+++ b/Tests/test_image.py
@@ -182,6 +182,18 @@ class TestImage:
         with Image.open(out) as reloaded:
             assert_image_equal(im, reloaded)
 
+    def test_dry_run(self, tmp_path: Path) -> None:
+        out = tmp_path / "temp.png"
+        im = hopper()
+        im.save(out, dry_run=True)
+        assert not out.exists()
+
+        # errors that would occur during saving are still raised
+        with hopper("RGBA") as im:
+            with pytest.raises(ValueError):
+                im.save(tmp_path / "temp.pcx", dry_run=True)
+        assert not (tmp_path / "temp.pcx").exists()
+
     def test_pathlib(self, tmp_path: Path) -> None:
         with Image.open(Path("Tests/images/multipage-mmap.tiff")) as im:
             assert im.mode == "P"
diff --git a/src/PIL/Image.py b/src/PIL/Image.py
index 233df592c..17bd0e7c0 100644
--- a/src/PIL/Image.py
+++ b/src/PIL/Image.py
@@ -2505,6 +2505,9 @@ class Image:
              red = Image.new("RGB", (1, 1), "#f00")
              red.encoderinfo = {"xmp": b"test"}
              black.save("out.mpo", save_all=True, append_images=[red])
+        :param dry_run: If true, the image will be processed for saving, so that
+           any errors that would occur during saving are raised, but the result
+           will not actually be written to disk.
         :returns: None
         :exception ValueError: If the output format could not be determined
            from the file name.  Use the format option to solve this.
@@ -2544,6 +2547,7 @@ class Image:
         self._ensure_mutable()
 
         save_all = params.pop("save_all", None)
+        dry_run = params.pop("dry_run", False)
         self.encoderinfo = {**getattr(self, "encoderinfo", {}), **params}
         self.encoderconfig: tuple[Any, ...] = ()
 
@@ -2559,7 +2563,10 @@ class Image:
             save_handler = SAVE[format.upper()]
 
         created = False
-        if open_fp:
+        if dry_run:
+            open_fp = False
+            fp = io.BytesIO()
+        elif open_fp:
             created = not os.path.exists(filename)
             if params.get("append", False):
                 # Open also for reading ("+"), because TIFF save_all
