diff --git a/CHANGES.rst b/CHANGES.rst
index 2db769f..101ae4c 100644
--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -62,6 +62,8 @@ Unreleased
 -   Contexts created during shell completion are closed properly, fixing
     ``ResourceWarning``s when using ``click.File``. :issue:`2644` :pr:`2800`
     :pr:`2767`
+-   ``click.edit()`` now supports editing multiple files at once by passing
+    an iterable of filenames to ``filename``.
 
 
 Version 8.1.8
diff --git a/src/click/_termui_impl.py b/src/click/_termui_impl.py
index 7b97bfb..99ff09e 100644
--- a/src/click/_termui_impl.py
+++ b/src/click/_termui_impl.py
@@ -505,7 +505,7 @@ class Editor:
                 return editor
         return "vi"
 
-    def edit_file(self, filename: str) -> None:
+    def edit_files(self, filenames: cabc.Iterable[str]) -> None:
         import subprocess
 
         editor = self.get_editor()
@@ -515,8 +515,9 @@ class Editor:
             environ = os.environ.copy()
             environ.update(self.env)
 
+        filenames = " ".join(f'"{filename}"' for filename in filenames)
         try:
-            c = subprocess.Popen(f'{editor} "{filename}"', env=environ, shell=True)
+            c = subprocess.Popen(f"{editor} {filenames}", env=environ, shell=True)
             exit_code = c.wait()
             if exit_code != 0:
                 raise ClickException(
@@ -559,7 +560,7 @@ class Editor:
             # recorded, so get the new recorded value.
             timestamp = os.path.getmtime(name)
 
-            self.edit_file(name)
+            self.edit_files((name,))
 
             if self.require_save and os.path.getmtime(name) == timestamp:
                 return None
diff --git a/src/click/termui.py b/src/click/termui.py
index e14e670..36bf6fa 100644
--- a/src/click/termui.py
+++ b/src/click/termui.py
@@ -643,13 +643,36 @@ def secho(
     return echo(message, file=file, nl=nl, err=err, color=color)
 
 
+@t.overload
 def edit(
     text: t.AnyStr | None = None,
     editor: str | None = None,
     env: cabc.Mapping[str, str] | None = None,
     require_save: bool = True,
     extension: str = ".txt",
-    filename: str | None = None,
+    filename: None = None,
+) -> t.AnyStr | None: ...
+
+
+@t.overload
+def edit(
+    text: t.AnyStr | None = None,
+    editor: str | None = None,
+    env: cabc.Mapping[str, str] | None = None,
+    require_save: bool = True,
+    extension: str = ".txt",
+    *,
+    filename: str | cabc.Iterable[str],
+) -> None: ...
+
+
+def edit(
+    text: t.AnyStr | None = None,
+    editor: str | None = None,
+    env: cabc.Mapping[str, str] | None = None,
+    require_save: bool = True,
+    extension: str = ".txt",
+    filename: str | cabc.Iterable[str] | None = None,
 ) -> t.AnyStr | None:
     r"""Edits the given text in the defined editor.  If an editor is given
     (should be the full path to the executable but the regular operating
@@ -676,7 +699,14 @@ def edit(
                       highlighting.
     :param filename: if provided it will edit this file instead of the
                      provided text contents.  It will not use a temporary
-                     file as an indirection in that case.
+                     file as an indirection in that case.  If the
+                     environment variable of ``EDITOR`` supports it, ``str``
+                     or ``Iterable[str]`` will allow for the specifying of
+                     multiple files at once.
+
+    .. versionchanged:: 8.2.0
+        ``filename`` now supports an iterable of filenames to edit multiple
+        files at once.
     """
     from ._termui_impl import Editor
 
@@ -685,7 +715,12 @@ def edit(
     if filename is None:
         return ed.edit(text)
 
-    ed.edit_file(filename)
+    if isinstance(filename, str):
+        filenames: cabc.Iterable[str] = (filename,)
+    else:
+        filenames = filename
+
+    ed.edit_files(filenames=filenames)
     return None
 
 
