diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..4cf4b0e 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -67,6 +67,7 @@ __all__ = (
     'IsInstance',
     # other
     'FunctionCheck',
+    'IsHash',
     'IsJson',
     'IsUUID',
     # strings
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..c61aaad 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,4 +1,5 @@
 import json
+import string
 from typing import Any, Callable, TypeVar, overload
 from uuid import UUID
 
@@ -145,3 +146,57 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+HashTypes = Literal['md5', 'sha-1', 'sha-224', 'sha-256', 'sha-384', 'sha-512']
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash of the given hash type, using a simple length and allowed
+    characters check.
+    """
+
+    def __init__(self, hash_type: HashTypes):
+        """
+        Args:
+            hash_type: The hash type to check. Must be specified, one of `'md5'`, `'sha-1'`, `'sha-224'`,
+                `'sha-256'`, `'sha-384'` or `'sha-512'`.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert 'f1e069787ece74531d112559945c6871' == IsHash('md5')
+        assert b'f1e069787ece74531d112559945c6871' == IsHash('md5')
+        assert 'f1e069787ece74531d112559945c6871' != IsHash('sha-256')
+        assert 'F1E069787ECE74531D112559945C6871' == IsHash('md5')
+        assert '40bd001563085fc35165329ea1ff5c5ecbdbbeef' == IsHash('sha-1')
+        assert 'a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3' == IsHash(
+            'sha-256'
+        )
+        ```
+        """
+        allowed_hashes = HashTypes.__args__  # type: ignore[attr-defined]
+        if hash_type not in allowed_hashes:
+            raise ValueError(f'Hash type must be one of the following values: {", ".join(allowed_hashes)}')
+
+        self.hash_type = hash_type
+        super().__init__(hash_type)
+
+    def equals(self, other: Any) -> bool:
+        if isinstance(other, str):
+            s = other
+        elif isinstance(other, (bytes, bytearray)):
+            s = other.decode()
+        else:
+            return False
+        hash_type_lengths = {
+            'md5': 32,
+            'sha-1': 40,
+            'sha-224': 56,
+            'sha-256': 64,
+            'sha-384': 96,
+            'sha-512': 128,
+        }
+        length = hash_type_lengths[self.hash_type]
+        return len(s) == length and all(c in string.hexdigits for c in s)
