diff --git a/src/jinja2/ext.py b/src/jinja2/ext.py
index d21b83aa..25e51675 100644
--- a/src/jinja2/ext.py
+++ b/src/jinja2/ext.py
@@ -47,6 +47,8 @@ GETTEXT_FUNCTIONS: t.Tuple[str, ...] = (
     "ngettext",
     "pgettext",
     "npgettext",
+    "dgettext",
+    "dngettext",
 )
 _ws_re = re.compile(r"\s*\n\s*")
 
@@ -242,6 +244,42 @@ def _make_new_npgettext(
     return npgettext
 
 
+def _make_new_dgettext(func: t.Callable[[str, str], str]) -> t.Callable[..., str]:
+    @pass_context
+    def dgettext(
+        __context: Context, __domain: str, __string: str, **variables: t.Any
+    ) -> str:
+        rv = __context.call(func, __domain, __string)
+        if __context.eval_ctx.autoescape:
+            rv = Markup(rv)
+        # Always treat as a format string, see gettext comment above.
+        return rv % variables  # type: ignore
+
+    return dgettext
+
+
+def _make_new_dngettext(
+    func: t.Callable[[str, str, str, int], str]
+) -> t.Callable[..., str]:
+    @pass_context
+    def dngettext(
+        __context: Context,
+        __domain: str,
+        __singular: str,
+        __plural: str,
+        __num: int,
+        **variables: t.Any,
+    ) -> str:
+        variables.setdefault("num", __num)
+        rv = __context.call(func, __domain, __singular, __plural, __num)
+        if __context.eval_ctx.autoescape:
+            rv = Markup(rv)
+        # Always treat as a format string, see gettext comment above.
+        return rv % variables  # type: ignore
+
+    return dngettext
+
+
 class InternationalizationExtension(Extension):
     """This extension adds gettext support to Jinja."""
 
@@ -280,8 +318,16 @@ class InternationalizationExtension(Extension):
 
         pgettext = getattr(translations, "pgettext", None)
         npgettext = getattr(translations, "npgettext", None)
+        dgettext = getattr(translations, "dgettext", None)
+        dngettext = getattr(translations, "dngettext", None)
         self._install_callables(
-            gettext, ngettext, newstyle=newstyle, pgettext=pgettext, npgettext=npgettext
+            gettext,
+            ngettext,
+            newstyle=newstyle,
+            pgettext=pgettext,
+            npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _install_null(self, newstyle: t.Optional[bool] = None) -> None:
@@ -319,6 +365,8 @@ class InternationalizationExtension(Extension):
         newstyle: t.Optional[bool] = None,
         pgettext: t.Optional[t.Callable[[str, str], str]] = None,
         npgettext: t.Optional[t.Callable[[str, str, str, int], str]] = None,
+        dgettext: t.Optional[t.Callable[[str, str], str]] = None,
+        dngettext: t.Optional[t.Callable[[str, str, str, int], str]] = None,
     ) -> None:
         if newstyle is not None:
             self.environment.newstyle_gettext = newstyle  # type: ignore
@@ -332,12 +380,71 @@ class InternationalizationExtension(Extension):
             if npgettext is not None:
                 npgettext = _make_new_npgettext(npgettext)
 
+            if dgettext is not None:
+                dgettext = _make_new_dgettext(dgettext)
+            else:
+                _gettext = gettext
+
+                @pass_context
+                def dgettext(
+                    __context: Context,
+                    __domain: str,
+                    __string: str,
+                    **variables: t.Any,
+                ) -> str:
+                    return __context.call(_gettext, __string, **variables)
+
+            if dngettext is not None:
+                dngettext = _make_new_dngettext(dngettext)
+            else:
+                _ngettext = ngettext
+
+                @pass_context
+                def dngettext(
+                    __context: Context,
+                    __domain: str,
+                    __singular: str,
+                    __plural: str,
+                    __num: int,
+                    **variables: t.Any,
+                ) -> str:
+                    return __context.call(
+                        _ngettext, __singular, __plural, __num, **variables
+                    )
+
+        else:
+            if dgettext is None:
+                _gettext = gettext
+
+                def dgettext(__domain: str, __string: str) -> str:
+                    return _gettext(__string)
+
+            if dngettext is None:
+                _ngettext = ngettext
+
+                def dngettext(
+                    __domain: str, __singular: str, __plural: str, __num: int
+                ) -> str:
+                    return _ngettext(__singular, __plural, __num)
+
         self.environment.globals.update(
-            gettext=gettext, ngettext=ngettext, pgettext=pgettext, npgettext=npgettext
+            gettext=gettext,
+            ngettext=ngettext,
+            pgettext=pgettext,
+            npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _uninstall(self, translations: "_SupportedTranslations") -> None:
-        for key in ("gettext", "ngettext", "pgettext", "npgettext"):
+        for key in (
+            "gettext",
+            "ngettext",
+            "pgettext",
+            "npgettext",
+            "dgettext",
+            "dngettext",
+        ):
             self.environment.globals.pop(key, None)
 
     def _extract(
@@ -363,6 +470,7 @@ class InternationalizationExtension(Extension):
         plural_expr_assignment: t.Optional[nodes.Assign] = None
         variables: t.Dict[str, nodes.Expr] = {}
         trimmed = None
+        domain: t.Optional[str] = None
         while parser.stream.current.type != "block_end":
             if variables:
                 parser.stream.expect("comma")
@@ -379,6 +487,34 @@ class InternationalizationExtension(Extension):
                     exc=TemplateAssertionError,
                 )
 
+            if token.value == "domain":
+                if domain is not None:
+                    parser.fail(
+                        "translation domain specified twice.",
+                        token.lineno,
+                        exc=TemplateAssertionError,
+                    )
+                if parser.stream.current.type != "assign":
+                    parser.fail(
+                        "the translation domain must be a string literal"
+                        ' constant, e.g. domain="mydomain".',
+                        token.lineno,
+                        exc=TemplateAssertionError,
+                    )
+                next(parser.stream)
+                domain_expr = parser.parse_expression()
+                if not isinstance(domain_expr, nodes.Const) or not isinstance(
+                    domain_expr.value, str
+                ):
+                    parser.fail(
+                        "the translation domain must be a string literal"
+                        " constant, not a variable or expression.",
+                        domain_expr.lineno,
+                        exc=TemplateAssertionError,
+                    )
+                domain = domain_expr.value
+                continue
+
             # expressions
             if parser.stream.current.type == "assign":
                 next(parser.stream)
@@ -459,6 +595,7 @@ class InternationalizationExtension(Extension):
             plural_expr,
             bool(referenced),
             num_called_num and have_plural,
+            domain,
         )
         node.set_lineno(lineno)
         if plural_expr_assignment is not None:
@@ -514,6 +651,7 @@ class InternationalizationExtension(Extension):
         plural_expr: t.Optional[nodes.Expr],
         vars_referenced: bool,
         num_called_num: bool,
+        domain: t.Optional[str] = None,
     ) -> nodes.Output:
         """Generates a useful node from the data provided."""
         newstyle = self.environment.newstyle_gettext  # type: ignore
@@ -528,15 +666,35 @@ class InternationalizationExtension(Extension):
 
         # singular only:
         if plural_expr is None:
-            gettext = nodes.Name("gettext", "load")
-            node = nodes.Call(gettext, [nodes.Const(singular)], [], None, None)
+            if domain is not None:
+                func_name = "dgettext"
+                func_args: t.List[nodes.Expr] = [
+                    nodes.Const(domain),
+                    nodes.Const(singular),
+                ]
+            else:
+                func_name = "gettext"
+                func_args = [nodes.Const(singular)]
+            gettext = nodes.Name(func_name, "load")
+            node = nodes.Call(gettext, func_args, [], None, None)
 
         # singular and plural
         else:
-            ngettext = nodes.Name("ngettext", "load")
+            if domain is not None:
+                func_name = "dngettext"
+                func_args = [
+                    nodes.Const(domain),
+                    nodes.Const(singular),
+                    nodes.Const(plural),
+                    plural_expr,
+                ]
+            else:
+                func_name = "ngettext"
+                func_args = [nodes.Const(singular), nodes.Const(plural), plural_expr]
+            ngettext = nodes.Name(func_name, "load")
             node = nodes.Call(
                 ngettext,
-                [nodes.Const(singular), nodes.Const(plural), plural_expr],
+                func_args,
                 [],
                 None,
                 None,
