diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..2a2aa1e 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,4 +1,5 @@
 import json
+import re
 from typing import Any, Callable, TypeVar, overload
 from uuid import UUID
 
@@ -120,6 +121,51 @@ class IsJson(DirtyEquals[JsonType]):
             return False
 
 
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash of a specific algorithm.
+    """
+
+    hash_lengths = {
+        'md5': 32,
+        'sha1': 40,
+        'sha224': 56,
+        'sha256': 64,
+        'sha384': 96,
+        'sha512': 128,
+    }
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm to use for the hash, currently supports 'md5', 'sha1', 'sha224',
+                'sha256', 'sha384' and 'sha512'.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '5eb63bbbe01eeed093cb22bb8f5acdc3' == IsHash('md5')
+        assert '2aae6c35c94fcfb415dbe95f408b9ce91ee846ed' == IsHash('sha1')
+        assert '9e763e6ff77103cccc51706c3d4f0fa4a1a26fb0aca0b23e463cd6828d8b3e34' != IsHash('sha1')
+        assert 'not a hash' != IsHash('sha256')
+        ```
+        """
+        self.algorithm = algorithm.lower()
+        if self.algorithm not in self.hash_lengths:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        self.hash_length = self.hash_lengths[self.algorithm]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        if len(other) != self.hash_length:
+            return False
+
+        return bool(re.fullmatch('[0-9a-fA-F]+', other))
+
+
 class FunctionCheck(DirtyEquals[Any]):
     """
     Use a function to check if a value "equals" whatever you want to check
