diff --git a/src/PIL/Image.py b/src/PIL/Image.py
index 233df592c..899543857 100644
--- a/src/PIL/Image.py
+++ b/src/PIL/Image.py
@@ -2467,7 +2467,13 @@ class Image:
         )
 
     def save(
-        self, fp: StrOrBytesPath | IO[bytes], format: str | None = None, **params: Any
+        self,
+        fp: StrOrBytesPath | IO[bytes],
+        format: str | None = None,
+        *,
+        preserve_readonly: bool = False,
+        dry_run: bool = False,
+        **params: Any,
     ) -> None:
         """
         Saves this image under the given filename.  If no format is
@@ -2490,6 +2496,14 @@ class Image:
            format to use is determined from the filename extension.
            If a file object was used instead of a filename, this
            parameter should always be used.
+        :param preserve_readonly: If true, the image's readonly state will not be
+           reset before saving, so that e.g. images created with
+           :py:meth:`~PIL.Image.frombuffer` will continue to reflect changes made
+           to the underlying buffer. Defaults to false, which ensures the image is
+           made mutable before saving.
+        :param dry_run: If true, the image is fully processed for saving, but the
+           result is not written to disk. This can be used to validate that an
+           image can be saved without the overhead of the file write.
         :param params: Extra parameters to the image writer. These can also be
            set on the image itself through ``encoderinfo``. This is useful when
            saving multiple images::
@@ -2541,7 +2555,10 @@ class Image:
                 raise ValueError(msg) from e
 
         # may mutate self!
-        self._ensure_mutable()
+        if preserve_readonly:
+            self.load()
+        else:
+            self._ensure_mutable()
 
         save_all = params.pop("save_all", None)
         self.encoderinfo = {**getattr(self, "encoderinfo", {}), **params}
@@ -2559,7 +2576,10 @@ class Image:
             save_handler = SAVE[format.upper()]
 
         created = False
-        if open_fp:
+        if dry_run:
+            fp = io.BytesIO()
+            open_fp = False
+        elif open_fp:
             created = not os.path.exists(filename)
             if params.get("append", False):
                 # Open also for reading ("+"), because TIFF save_all
