diff --git a/src/jinja2/filters.py b/src/jinja2/filters.py
index 80ea6504..85b7dd61 100644
--- a/src/jinja2/filters.py
+++ b/src/jinja2/filters.py
@@ -1163,6 +1163,7 @@ def sync_do_groupby(
     value: "t.Iterable[V]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    reverse: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     """Group a sequence of objects by an attribute using Python's
     :func:`itertools.groupby`. The attribute can use dot notation for
@@ -1203,6 +1204,18 @@ def sync_do_groupby(
           <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
         {% endfor %}</ul>
 
+    You can also use ``reverse=True`` to sort the groups in descending
+    order.
+
+    .. sourcecode:: jinja
+
+        <ul>{% for city, items in users|groupby("city", reverse=true) %}
+          <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
+        {% endfor %}</ul>
+
+    .. versionchanged:: 3.1
+        Added the ``reverse`` parameter.
+
     .. versionchanged:: 3.0
         Added the ``default`` parameter.
 
@@ -1212,7 +1225,7 @@ def sync_do_groupby(
     expr = make_attrgetter(environment, attribute, default=default)
     return [
         _GroupTuple(key, list(values))
-        for key, values in groupby(sorted(value, key=expr), expr)
+        for key, values in groupby(sorted(value, key=expr, reverse=reverse), expr)
     ]
 
 
@@ -1222,11 +1235,14 @@ async def do_groupby(
     value: "t.Union[t.AsyncIterable[V], t.Iterable[V]]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    reverse: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     expr = make_attrgetter(environment, attribute, default=default)
     return [
         _GroupTuple(key, await auto_to_list(values))
-        for key, values in groupby(sorted(await auto_to_list(value), key=expr), expr)
+        for key, values in groupby(
+            sorted(await auto_to_list(value), key=expr, reverse=reverse), expr
+        )
     ]
 
 
