diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..5f41e1e 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -120,6 +120,51 @@ class IsJson(DirtyEquals[JsonType]):
             return False
 
 
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, for a given hashing algorithm.
+    """
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm to check against, one of `'md5'`, `'sha1'`, `'sha224'`, `'sha256'`,
+                `'sha384'`, `'sha512'`.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '9e107d9d372bb6826bd81d3542a419d6' == IsHash('md5')
+        assert 'aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d' == IsHash('sha1')
+        assert 'not a hash' != IsHash('md5')
+        ```
+        """
+        self.algorithm = algorithm
+        self.hash_length = {
+            'md5': 32,
+            'sha1': 40,
+            'sha224': 56,
+            'sha256': 64,
+            'sha384': 96,
+            'sha512': 128,
+        }[algorithm.lower()]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        if len(other) != self.hash_length:
+            return False
+
+        try:
+            int(other, 16)
+        except ValueError:
+            return False
+
+        return True
+
+
 class FunctionCheck(DirtyEquals[Any]):
     """
     Use a function to check if a value "equals" whatever you want to check
