diff --git a/src/jinja2/filters.py b/src/jinja2/filters.py
index 80ea6504..1823e00f 100644
--- a/src/jinja2/filters.py
+++ b/src/jinja2/filters.py
@@ -1163,6 +1163,7 @@ def sync_do_groupby(
     value: "t.Iterable[V]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     """Group a sequence of objects by an attribute using Python's
     :func:`itertools.groupby`. The attribute can use dot notation for
@@ -1203,6 +1204,17 @@ def sync_do_groupby(
           <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
         {% endfor %}</ul>
 
+    Like the :func:`groupby` filter, sorting and grouping is case
+    insensitive by default. The ``key`` for each group will have the
+    case of the first item in that group of values. For example, if a
+    list of users has cities ``["CA", "NY", "ca"]``, the "CA" and "ca"
+    values will be grouped together, and the ``grouper`` will be ``"CA"``
+    for that group. Pass ``case_sensitive=True`` to group values by
+    distinct case.
+
+    .. versionchanged:: 3.1
+        Added the ``case_sensitive`` parameter.
+
     .. versionchanged:: 3.0
         Added the ``default`` parameter.
 
@@ -1210,10 +1222,21 @@ def sync_do_groupby(
         The attribute supports dot notation for nested access.
     """
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, list(values))
-        for key, values in groupby(sorted(value, key=expr), expr)
-    ]
+
+    if case_sensitive:
+        key_func = expr
+    else:
+        key_func = make_attrgetter(
+            environment, attribute, postprocess=ignore_case, default=default
+        )
+
+    out = []
+
+    for _, values in groupby(sorted(value, key=key_func), key_func):
+        values = list(values)
+        out.append(_GroupTuple(expr(values[0]), values))
+
+    return out
 
 
 @async_variant(sync_do_groupby)  # type: ignore
@@ -1222,12 +1245,26 @@ async def do_groupby(
     value: "t.Union[t.AsyncIterable[V], t.Iterable[V]]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, await auto_to_list(values))
-        for key, values in groupby(sorted(await auto_to_list(value), key=expr), expr)
-    ]
+
+    if case_sensitive:
+        key_func = expr
+    else:
+        key_func = make_attrgetter(
+            environment, attribute, postprocess=ignore_case, default=default
+        )
+
+    out = []
+
+    for _, values in groupby(
+        sorted(await auto_to_list(value), key=key_func), key_func
+    ):
+        values = list(values)
+        out.append(_GroupTuple(expr(values[0]), values))
+
+    return out
 
 
 @pass_environment
