diff --git a/llama-index-core/llama_index/core/base/llms/types.py b/llama-index-core/llama_index/core/base/llms/types.py
index 00e30cf23..2b8f85543 100644
--- a/llama-index-core/llama_index/core/base/llms/types.py
+++ b/llama-index-core/llama_index/core/base/llms/types.py
@@ -108,6 +108,17 @@ class ImageBlock(BaseModel):
         else:
             raise ValueError("No image found in the chat message!")
 
+    def get_image_mimetype(self) -> str | None:
+        """Return the image mimetype, guessing and caching it if not already set."""
+        if self.image_mimetype:
+            return self.image_mimetype
+
+        if self.image:
+            guess = filetype.guess(self.image)
+            self.image_mimetype = guess.mime if guess else None
+
+        return self.image_mimetype
+
 
 ContentBlock = Annotated[
     Union[TextBlock, ImageBlock], Field(discriminator="block_type")
