diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..73f7e65 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsEmail, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -67,6 +67,7 @@ __all__ = (
     'IsInstance',
     # other
     'FunctionCheck',
+    'IsEmail',
     'IsJson',
     'IsUUID',
     # strings
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..3729416 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,5 +1,6 @@
 import json
-from typing import Any, Callable, TypeVar, overload
+import re
+from typing import Any, Callable, Optional, TypeVar, overload
 from uuid import UUID
 
 from ._base import DirtyEquals
@@ -51,6 +52,44 @@ class IsUUID(DirtyEquals[UUID]):
             return True
 
 
+class IsEmail(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid email address, optionally checking the email's domain.
+    """
+
+    _EMAIL_REGEX = re.compile(r'^[a-zA-Z0-9_.+-]+@([a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+)$')
+
+    def __init__(self, domain: Optional[str] = None):
+        """
+        Args:
+            domain: The domain of the email address to check, if omitted, any domain is accepted.
+
+        ```py title="IsEmail"
+        from dirty_equals import IsEmail
+
+        assert 'sample@example.com' == IsEmail
+        assert 'sample@example.com' == IsEmail(domain='example.com')
+        assert 'sample@example.com' != IsEmail(domain='other.com')
+        assert 'not-an-email' != IsEmail
+        ```
+        """
+        self.domain = domain
+        super().__init__(domain or plain_repr('*'))
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        match = self._EMAIL_REGEX.match(other)
+        if not match:
+            return False
+
+        if self.domain is not None and match.group(1) != self.domain:
+            return False
+
+        return True
+
+
 AnyJson = object
 JsonType = TypeVar('JsonType', AnyJson, Any)
 
