diff --git a/mux.go b/mux.go
index 56fa4d2..5d4b198 100644
--- a/mux.go
+++ b/mux.go
@@ -454,6 +454,7 @@ func (mx *Mux) routeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	// Find the route
 	if _, _, h := mx.tree.FindRoute(rctx, method, routePath); h != nil {
+		setPathValues(r, rctx)
 		h.ServeHTTP(w, r)
 		return
 	}
diff --git a/path_value.go b/path_value.go
new file mode 100644
index 0000000..3584191
--- /dev/null
+++ b/path_value.go
@@ -0,0 +1,19 @@
+//go:build go1.22
+// +build go1.22
+
+package chi
+
+import "net/http"
+
+// setPathValues sets the path values for the request, as detailed
+// in the net/http package for the (*http.Request) PathValue method,
+// added in Go 1.22.
+func setPathValues(r *http.Request, routeCtx *Context) {
+	for i, key := range routeCtx.URLParams.Keys {
+		if key == "*" || key == "" {
+			continue
+		}
+		value := routeCtx.URLParams.Values[i]
+		r.SetPathValue(key, value)
+	}
+}
diff --git a/path_value_fallback.go b/path_value_fallback.go
new file mode 100644
index 0000000..6fd644f
--- /dev/null
+++ b/path_value_fallback.go
@@ -0,0 +1,10 @@
+//go:build !go1.22
+// +build !go1.22
+
+package chi
+
+import "net/http"
+
+// setPathValues does nothing in versions of Go before 1.22, since
+// (*http.Request).SetPathValue was added in Go 1.22.
+func setPathValues(r *http.Request, routeCtx *Context) {}
