diff --git a/src/jinja2/filters.py b/src/jinja2/filters.py
index 80ea6504..6200ae47 100644
--- a/src/jinja2/filters.py
+++ b/src/jinja2/filters.py
@@ -1157,12 +1157,28 @@ class _GroupTuple(t.NamedTuple):
         return tuple.__str__(self)
 
 
+class _GroupTupleWithCount(t.NamedTuple):
+    grouper: t.Any
+    list: t.List
+    count: int
+
+    # Use the regular tuple repr to hide this subclass if users print
+    # out the value during debugging.
+    def __repr__(self) -> str:
+        return tuple.__repr__(self)
+
+    def __str__(self) -> str:
+        return tuple.__str__(self)
+
+
 @pass_environment
 def sync_do_groupby(
     environment: "Environment",
     value: "t.Iterable[V]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
+    with_counts: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     """Group a sequence of objects by an attribute using Python's
     :func:`itertools.groupby`. The attribute can use dot notation for
@@ -1203,17 +1219,54 @@ def sync_do_groupby(
           <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
         {% endfor %}</ul>
 
+    Like the :func:`~jinja-filters.sort` filter, sorting and grouping is
+    case-insensitive by default. The ``key`` for each group will have
+    the case of the first item in that group of values. For example, if
+    a list of users has cities ``["CA", "NY", "ca"]``, the "CA" group
+    will have two values. This can be disabled by passing
+    ``case_sensitive=True``.
+
+    You can also get the number of items in each group without needing
+    to consume the whole group first by passing ``with_counts=True``,
+    which adds a third ``count`` item to each returned tuple.
+
+    .. sourcecode:: jinja
+
+        <ul>{% for city, items, count in users|groupby("city", with_counts=True) %}
+          <li>{{ city }} ({{ count }})</li>
+        {% endfor %}</ul>
+
+    .. versionchanged:: 3.1
+        Added the ``with_counts`` parameter.
+
+    .. versionchanged:: 3.1
+        Added the ``case_sensitive`` parameter.
+
     .. versionchanged:: 3.0
         Added the ``default`` parameter.
 
     .. versionchanged:: 2.6
         The attribute supports dot notation for nested access.
     """
-    expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, list(values))
-        for key, values in groupby(sorted(value, key=expr), expr)
-    ]
+    expr = make_attrgetter(
+        environment,
+        attribute,
+        postprocess=ignore_case if not case_sensitive else None,
+        default=default,
+    )
+    raw_expr = make_attrgetter(environment, attribute, default=default)
+    out = []
+
+    for _, g in groupby(sorted(value, key=expr), expr):
+        values = list(g)
+        grouper = raw_expr(values[0])
+
+        if with_counts:
+            out.append(_GroupTupleWithCount(grouper, values, len(values)))
+        else:
+            out.append(_GroupTuple(grouper, values))
+
+    return out
 
 
 @async_variant(sync_do_groupby)  # type: ignore
@@ -1222,12 +1275,28 @@ async def do_groupby(
     value: "t.Union[t.AsyncIterable[V], t.Iterable[V]]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
+    with_counts: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
-    expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, await auto_to_list(values))
-        for key, values in groupby(sorted(await auto_to_list(value), key=expr), expr)
-    ]
+    expr = make_attrgetter(
+        environment,
+        attribute,
+        postprocess=ignore_case if not case_sensitive else None,
+        default=default,
+    )
+    raw_expr = make_attrgetter(environment, attribute, default=default)
+    out = []
+
+    for _, g in groupby(sorted(await auto_to_list(value), key=expr), expr):
+        values = list(g)
+        grouper = raw_expr(values[0])
+
+        if with_counts:
+            out.append(_GroupTupleWithCount(grouper, values, len(values)))
+        else:
+            out.append(_GroupTuple(grouper, values))
+
+    return out
 
 
 @pass_environment
