diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..bf621f5 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,5 +1,6 @@
 import json
-from typing import Any, Callable, TypeVar, overload
+import re
+from typing import Any, Callable, Dict, TypeVar, overload
 from uuid import UUID
 
 from ._base import DirtyEquals
@@ -145,3 +146,46 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash of a given algorithm, this is done by checking the type, length
+    and character set of the value.
+    """
+
+    hash_lengths: Dict[str, int] = {
+        'md5': 32,
+        'sha1': 40,
+        'sha224': 56,
+        'sha256': 64,
+        'sha384': 96,
+        'sha512': 128,
+    }
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm to use for the hash, must be one of
+                `'md5'`, `'sha1'`, `'sha224'`, `'sha256'`, `'sha384'`, `'sha512'`.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert 'd41d8cd98f00b204e9800998ecf8427e' == IsHash('md5')
+        assert 'da39a3ee5e6b4b0d3255bfef95601890afd80709' == IsHash('sha1')
+        assert '2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae' == IsHash('sha256')
+        assert 'not-a-valid-hash' != IsHash('sha256')
+        ```
+        """
+        algorithm_lower = algorithm.lower()
+        if algorithm_lower not in self.hash_lengths:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        self.algorithm = algorithm
+        self.hash_length = self.hash_lengths[algorithm_lower]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        return len(other) == self.hash_length and bool(re.fullmatch('[0-9a-fA-F]+', other))
