diff --git a/src/PIL/Image.py b/src/PIL/Image.py
index 233df592c..188c817db 100644
--- a/src/PIL/Image.py
+++ b/src/PIL/Image.py
@@ -2540,8 +2540,11 @@ class Image:
                 msg = f"unknown file extension: {ext}"
                 raise ValueError(msg) from e
 
-        # may mutate self!
-        self._ensure_mutable()
+        if filename == getattr(self, "filename", None):
+            # may mutate self!
+            self._ensure_mutable()
+        else:
+            self.load()
 
         save_all = params.pop("save_all", None)
         self.encoderinfo = {**getattr(self, "encoderinfo", {}), **params}
