diff --git a/src/jinja2/filters.py b/src/jinja2/filters.py
index 80ea6504..d213beec 100644
--- a/src/jinja2/filters.py
+++ b/src/jinja2/filters.py
@@ -1163,6 +1163,7 @@ def sync_do_groupby(
     value: "t.Iterable[V]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     """Group a sequence of objects by an attribute using Python's
     :func:`itertools.groupby`. The attribute can use dot notation for
@@ -1203,6 +1204,18 @@ def sync_do_groupby(
           <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
         {% endfor %}</ul>
 
+    Like the :func:`~jinja-filters.sort` filter, ``groupby`` is case
+    insensitive by default. The ``key`` for each group will have the
+    case of the first item in that group of values. For example, if a
+    list of users has cities ``["CA", "NY", "ca"]``, the "CA" group
+    will have two values. This can be disabled by passing
+    ``case_sensitive=True``.
+
+    .. versionchanged:: 3.1
+        Added the ``case_sensitive`` parameter. Sorting and grouping
+        is case-insensitive by default, matching other filters that
+        do comparisons.
+
     .. versionchanged:: 3.0
         Added the ``default`` parameter.
 
@@ -1210,10 +1223,19 @@ def sync_do_groupby(
         The attribute supports dot notation for nested access.
     """
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, list(values))
-        for key, values in groupby(sorted(value, key=expr), expr)
-    ]
+    sort_key = make_attrgetter(
+        environment,
+        attribute,
+        postprocess=None if case_sensitive else ignore_case,
+        default=default,
+    )
+    out = []
+
+    for _, values in groupby(sorted(value, key=sort_key), sort_key):
+        values = list(values)
+        out.append(_GroupTuple(expr(values[0]), values))
+
+    return out
 
 
 @async_variant(sync_do_groupby)  # type: ignore
@@ -1222,12 +1244,24 @@ async def do_groupby(
     value: "t.Union[t.AsyncIterable[V], t.Iterable[V]]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, await auto_to_list(values))
-        for key, values in groupby(sorted(await auto_to_list(value), key=expr), expr)
-    ]
+    sort_key = make_attrgetter(
+        environment,
+        attribute,
+        postprocess=None if case_sensitive else ignore_case,
+        default=default,
+    )
+    out = []
+
+    for _, values in groupby(
+        sorted(await auto_to_list(value), key=sort_key), sort_key
+    ):
+        values = await auto_to_list(values)
+        out.append(_GroupTuple(expr(values[0]), values))
+
+    return out
 
 
 @pass_environment
