diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..50ddd7c 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,5 +1,6 @@
 import json
-from typing import Any, Callable, TypeVar, overload
+import re
+from typing import Any, Callable, Dict, TypeVar, overload
 from uuid import UUID
 
 from ._base import DirtyEquals
@@ -145,3 +146,61 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+HashTypes = Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash-like string of a given algorithm.
+    """
+
+    hash_lengths: Dict[str, int] = {
+        'md5': 32,
+        'sha1': 40,
+        'sha224': 56,
+        'sha256': 64,
+        'sha384': 96,
+        'sha512': 128,
+    }
+
+    def __init__(self, algorithm: HashTypes):
+        """
+        Args:
+            algorithm: The algorithm to use for the hash, must be one of `'md5'`, `'sha1'`, `'sha224'`,
+                `'sha256'`, `'sha384'`, or `'sha512'`.
+
+        ```py title="IsHash"
+        import hashlib
+        from dirty_equals import IsHash
+
+        s = 'The quick brown fox jumps over the lazy dog'
+
+        md5 = hashlib.md5(s.encode()).hexdigest()
+        assert md5 == IsHash('md5')
+
+        sha1 = hashlib.sha1(s.encode()).hexdigest()
+        assert sha1 == IsHash('sha1')
+
+        sha256 = hashlib.sha256(s.encode()).hexdigest()
+        assert sha256 == IsHash('sha256')
+
+        assert 'not_a_hash' != IsHash('sha256')
+        ```
+        """
+        self._algorithm = algorithm.lower()
+        if self._algorithm not in self.hash_lengths:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+
+        self.hash_length = self.hash_lengths[self._algorithm]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        if len(other) != self.hash_length:
+            return False
+
+        return bool(re.fullmatch(r'[0-9a-fA-F]+', other))
