diff --git a/src/jinja2/loaders.py b/src/jinja2/loaders.py
index d7d9bd04..4b1e0dd8 100644
--- a/src/jinja2/loaders.py
+++ b/src/jinja2/loaders.py
@@ -170,6 +170,12 @@ class FileSystemLoader(BaseLoader):
     :param encoding: Use this encoding to read the text from template
         files.
     :param followlinks: Follow symbolic links in the path.
+    :param fallback_searchpath: A path, or list of paths, to search if
+        the template is not found in ``searchpath``. Primary paths in
+        ``searchpath`` always take precedence.
+
+    .. versionchanged:: 3.2
+        Added the ``fallback_searchpath`` parameter.
 
     .. versionchanged:: 2.8
         Added the ``followlinks`` parameter.
@@ -180,11 +186,23 @@ class FileSystemLoader(BaseLoader):
         searchpath: t.Union[str, os.PathLike, t.Sequence[t.Union[str, os.PathLike]]],
         encoding: str = "utf-8",
         followlinks: bool = False,
+        fallback_searchpath: t.Optional[
+            t.Union[str, os.PathLike, t.Sequence[t.Union[str, os.PathLike]]]
+        ] = None,
     ) -> None:
         if not isinstance(searchpath, abc.Iterable) or isinstance(searchpath, str):
             searchpath = [searchpath]
 
         self.searchpath = [os.fspath(p) for p in searchpath]
+
+        if fallback_searchpath is None:
+            fallback_searchpath = []
+        elif not isinstance(fallback_searchpath, abc.Iterable) or isinstance(
+            fallback_searchpath, str
+        ):
+            fallback_searchpath = [fallback_searchpath]
+
+        self.fallback_searchpath = [os.fspath(p) for p in fallback_searchpath]
         self.encoding = encoding
         self.followlinks = followlinks
 
@@ -192,7 +210,7 @@ class FileSystemLoader(BaseLoader):
         self, environment: "Environment", template: str
     ) -> t.Tuple[str, str, t.Callable[[], bool]]:
         pieces = split_template_path(template)
-        for searchpath in self.searchpath:
+        for searchpath in self.searchpath + self.fallback_searchpath:
             filename = os.path.join(searchpath, *pieces)
             f = open_if_exists(filename)
             if f is None:
@@ -215,7 +233,7 @@ class FileSystemLoader(BaseLoader):
 
     def list_templates(self) -> t.List[str]:
         found = set()
-        for searchpath in self.searchpath:
+        for searchpath in self.searchpath + self.fallback_searchpath:
             walk_dir = os.walk(searchpath, followlinks=self.followlinks)
             for dirpath, _, filenames in walk_dir:
                 for filename in filenames:
