diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..5d442f2 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,5 +1,6 @@
 import json
-from typing import Any, Callable, TypeVar, overload
+import re
+from typing import Any, Callable, Dict, TypeVar, overload
 from uuid import UUID
 
 from ._base import DirtyEquals
@@ -145,3 +146,47 @@ class FunctionCheck(DirtyEquals[Any]):
 
     def equals(self, other: Any) -> bool:
         return self.func(other)
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, tests that the value is a string representing a valid hash of the
+    given algorithm.
+    """
+
+    hash_length: Dict[str, int] = {
+        'md5': 32,
+        'sha1': 40,
+        'sha224': 56,
+        'sha256': 64,
+        'sha384': 96,
+        'sha512': 128,
+    }
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm which was used for encryption.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '5eb63bbbe01eeed093cb22bb8f5acdc3' == IsHash('md5')
+        assert 'aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d' == IsHash('sha1')
+        assert '2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae' != IsHash('sha256')
+        ```
+        """
+        self.algorithm = algorithm.lower()
+        if self.algorithm not in self.hash_length:
+            raise ValueError(f'Unknown algorithm {algorithm!r}')
+        self.length = self.hash_length[self.algorithm]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+
+        if len(other) != self.length:
+            return False
+
+        return bool(re.fullmatch('[0-9a-fA-F]+', other))
