diff --git a/mux.go b/mux.go
index 56fa4d2..950eaab 100644
--- a/mux.go
+++ b/mux.go
@@ -454,6 +454,7 @@ func (mx *Mux) routeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	// Find the route
 	if _, _, h := mx.tree.FindRoute(rctx, method, routePath); h != nil {
+		setPathValue(rctx, r)
 		h.ServeHTTP(w, r)
 		return
 	}
diff --git a/path_value.go b/path_value.go
new file mode 100644
index 0000000..98b506b
--- /dev/null
+++ b/path_value.go
@@ -0,0 +1,20 @@
+//go:build go1.22
+// +build go1.22
+
+package chi
+
+import "net/http"
+
+// setPathValue sets the path values in the Request value
+// based on the provided request context.
+//
+// It's a no-op for go < 1.22.
+func setPathValue(rctx *Context, r *http.Request) {
+	for i, key := range rctx.URLParams.Keys {
+		if key == "*" || key == "" {
+			continue
+		}
+		value := rctx.URLParams.Values[i]
+		r.SetPathValue(key, value)
+	}
+}
diff --git a/path_value_fallback.go b/path_value_fallback.go
new file mode 100644
index 0000000..9ab9037
--- /dev/null
+++ b/path_value_fallback.go
@@ -0,0 +1,12 @@
+//go:build !go1.22
+// +build !go1.22
+
+package chi
+
+import "net/http"
+
+// setPathValue sets the path values in the Request value
+// based on the provided request context.
+//
+// It's a no-op for go < 1.22.
+func setPathValue(rctx *Context, r *http.Request) {}
