diff --git a/outlines/types/__init__.py b/outlines/types/__init__.py
index 85af9adc..049bedcc 100644
--- a/outlines/types/__init__.py
+++ b/outlines/types/__init__.py
@@ -68,3 +68,6 @@ email = Regex(
 isbn = Regex(
     r"(?:ISBN(?:-1[03])?:? )?(?=[0-9X]{10}$|(?=(?:[0-9]+[- ]){3})[- 0-9X]{13}$|97[89][0-9]{10}$|(?=(?:[0-9]+[- ]){4})[- 0-9]{17}$)(?:97[89][- ]?)?[0-9]{1,5}[- ]?[0-9]+[- ]?[0-9]+[- ]?[0-9X]"
 )
+
+# Matches a SHA-256 hash represented as 64 hexadecimal characters
+hash_sha256 = Regex(r"[a-fA-F0-9]{64}")
