diff --git a/src/jinja2/ext.py b/src/jinja2/ext.py
index d21b83aa..ca7cfd51 100644
--- a/src/jinja2/ext.py
+++ b/src/jinja2/ext.py
@@ -242,6 +242,42 @@ def _make_new_npgettext(
     return npgettext
 
 
+def _make_new_dgettext(func: t.Callable[[str, str], str]) -> t.Callable[..., str]:
+    @pass_context
+    def dgettext(
+        __context: Context, __domain: str, __string: str, **variables: t.Any
+    ) -> str:
+        rv = __context.call(func, __domain, __string)
+        if __context.eval_ctx.autoescape:
+            rv = Markup(rv)
+        # Always treat as a format string, see gettext comment above.
+        return rv % variables  # type: ignore
+
+    return dgettext
+
+
+def _make_new_dngettext(
+    func: t.Callable[[str, str, str, int], str]
+) -> t.Callable[..., str]:
+    @pass_context
+    def dngettext(
+        __context: Context,
+        __domain: str,
+        __singular: str,
+        __plural: str,
+        __num: int,
+        **variables: t.Any,
+    ) -> str:
+        variables.setdefault("num", __num)
+        rv = __context.call(func, __domain, __singular, __plural, __num)
+        if __context.eval_ctx.autoescape:
+            rv = Markup(rv)
+        # Always treat as a format string, see gettext comment above.
+        return rv % variables  # type: ignore
+
+    return dngettext
+
+
 class InternationalizationExtension(Extension):
     """This extension adds gettext support to Jinja."""
 
@@ -280,8 +316,16 @@ class InternationalizationExtension(Extension):
 
         pgettext = getattr(translations, "pgettext", None)
         npgettext = getattr(translations, "npgettext", None)
+        dgettext = getattr(translations, "dgettext", None)
+        dngettext = getattr(translations, "dngettext", None)
         self._install_callables(
-            gettext, ngettext, newstyle=newstyle, pgettext=pgettext, npgettext=npgettext
+            gettext,
+            ngettext,
+            newstyle=newstyle,
+            pgettext=pgettext,
+            npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _install_null(self, newstyle: t.Optional[bool] = None) -> None:
@@ -304,12 +348,28 @@ class InternationalizationExtension(Extension):
             def npgettext(c: str, s: str, p: str, n: int) -> str:
                 return s if n == 1 else p
 
+        if hasattr(translations, "dgettext"):
+            dgettext = translations.dgettext  # type: ignore
+        else:
+
+            def dgettext(d: str, s: str) -> str:
+                return s
+
+        if hasattr(translations, "dngettext"):
+            dngettext = translations.dngettext  # type: ignore
+        else:
+
+            def dngettext(d: str, s: str, p: str, n: int) -> str:
+                return s if n == 1 else p
+
         self._install_callables(
             gettext=translations.gettext,
             ngettext=translations.ngettext,
             newstyle=newstyle,
             pgettext=pgettext,
             npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _install_callables(
@@ -319,6 +379,8 @@ class InternationalizationExtension(Extension):
         newstyle: t.Optional[bool] = None,
         pgettext: t.Optional[t.Callable[[str, str], str]] = None,
         npgettext: t.Optional[t.Callable[[str, str, str, int], str]] = None,
+        dgettext: t.Optional[t.Callable[[str, str], str]] = None,
+        dngettext: t.Optional[t.Callable[[str, str, str, int], str]] = None,
     ) -> None:
         if newstyle is not None:
             self.environment.newstyle_gettext = newstyle  # type: ignore
@@ -332,12 +394,30 @@ class InternationalizationExtension(Extension):
             if npgettext is not None:
                 npgettext = _make_new_npgettext(npgettext)
 
+            if dgettext is not None:
+                dgettext = _make_new_dgettext(dgettext)
+
+            if dngettext is not None:
+                dngettext = _make_new_dngettext(dngettext)
+
         self.environment.globals.update(
-            gettext=gettext, ngettext=ngettext, pgettext=pgettext, npgettext=npgettext
+            gettext=gettext,
+            ngettext=ngettext,
+            pgettext=pgettext,
+            npgettext=npgettext,
+            dgettext=dgettext,
+            dngettext=dngettext,
         )
 
     def _uninstall(self, translations: "_SupportedTranslations") -> None:
-        for key in ("gettext", "ngettext", "pgettext", "npgettext"):
+        for key in (
+            "gettext",
+            "ngettext",
+            "pgettext",
+            "npgettext",
+            "dgettext",
+            "dngettext",
+        ):
             self.environment.globals.pop(key, None)
 
     def _extract(
@@ -363,8 +443,9 @@ class InternationalizationExtension(Extension):
         plural_expr_assignment: t.Optional[nodes.Assign] = None
         variables: t.Dict[str, nodes.Expr] = {}
         trimmed = None
+        domain: t.Optional[str] = None
         while parser.stream.current.type != "block_end":
-            if variables:
+            if variables or domain is not None:
                 parser.stream.expect("comma")
 
             # skip colon for python compatibility
@@ -379,6 +460,31 @@ class InternationalizationExtension(Extension):
                     exc=TemplateAssertionError,
                 )
 
+            # the translation domain: domain="some_domain"
+            if token.value == "domain" and parser.stream.current.type == "assign":
+                next(parser.stream)
+                domain_token = parser.stream.current
+                domain_expr = parser.parse_expression()
+
+                if domain is not None:
+                    parser.fail(
+                        "translation domain specified twice",
+                        domain_token.lineno,
+                        exc=TemplateAssertionError,
+                    )
+
+                if not isinstance(domain_expr, nodes.Const) or not isinstance(
+                    domain_expr.value, str
+                ):
+                    parser.fail(
+                        "translation domain must be a string literal / constant",
+                        domain_token.lineno,
+                        exc=TemplateAssertionError,
+                    )
+
+                domain = domain_expr.value
+                continue
+
             # expressions
             if parser.stream.current.type == "assign":
                 next(parser.stream)
@@ -459,6 +565,7 @@ class InternationalizationExtension(Extension):
             plural_expr,
             bool(referenced),
             num_called_num and have_plural,
+            domain,
         )
         node.set_lineno(lineno)
         if plural_expr_assignment is not None:
@@ -514,10 +621,12 @@ class InternationalizationExtension(Extension):
         plural_expr: t.Optional[nodes.Expr],
         vars_referenced: bool,
         num_called_num: bool,
+        domain: t.Optional[str] = None,
     ) -> nodes.Output:
         """Generates a useful node from the data provided."""
         newstyle = self.environment.newstyle_gettext  # type: ignore
         node: nodes.Expr
+        args: t.List[nodes.Expr]
 
         # no variables referenced?  no need to escape for old style
         # gettext invocations only if there are vars.
@@ -528,19 +637,30 @@ class InternationalizationExtension(Extension):
 
         # singular only:
         if plural_expr is None:
-            gettext = nodes.Name("gettext", "load")
-            node = nodes.Call(gettext, [nodes.Const(singular)], [], None, None)
+            if domain is None:
+                func_name = "gettext"
+                args = [nodes.Const(singular)]
+            else:
+                func_name = "dgettext"
+                args = [nodes.Const(domain), nodes.Const(singular)]
+            func = nodes.Name(func_name, "load")
+            node = nodes.Call(func, args, [], None, None)
 
         # singular and plural
         else:
-            ngettext = nodes.Name("ngettext", "load")
-            node = nodes.Call(
-                ngettext,
-                [nodes.Const(singular), nodes.Const(plural), plural_expr],
-                [],
-                None,
-                None,
-            )
+            if domain is None:
+                func_name = "ngettext"
+                args = [nodes.Const(singular), nodes.Const(plural), plural_expr]
+            else:
+                func_name = "dngettext"
+                args = [
+                    nodes.Const(domain),
+                    nodes.Const(singular),
+                    nodes.Const(plural),
+                    plural_expr,
+                ]
+            func = nodes.Name(func_name, "load")
+            node = nodes.Call(func, args, [], None, None)
 
         # in case newstyle gettext is used, the method is powerful
         # enough to handle the variable expansion and autoescape
