diff --git a/src/click/shell_completion.py b/src/click/shell_completion.py
index 6fd9e54..ea811fc 100644
--- a/src/click/shell_completion.py
+++ b/src/click/shell_completion.py
@@ -544,8 +544,8 @@ def _resolve_context(
     :param args: List of complete args before the incomplete value.
     """
     ctx_args["resilient_parsing"] = True
-    ctx = cli.make_context(prog_name, args.copy(), **ctx_args)
-    args = ctx._protected_args + ctx.args
+    with cli.make_context(prog_name, args.copy(), **ctx_args) as ctx:
+        args = ctx._protected_args + ctx.args
 
     while args:
         command = ctx.command
@@ -557,8 +557,12 @@ def _resolve_context(
                 if cmd is None:
                     return ctx
 
-                ctx = cmd.make_context(name, args, parent=ctx, resilient_parsing=True)
-                args = ctx._protected_args + ctx.args
+                with cmd.make_context(
+                    name, args, parent=ctx, resilient_parsing=True
+                ) as sub_ctx:
+                    args = sub_ctx._protected_args + sub_ctx.args
+
+                ctx = sub_ctx
             else:
                 sub_ctx = ctx
 
@@ -568,15 +572,16 @@ def _resolve_context(
                     if cmd is None:
                         return ctx
 
-                    sub_ctx = cmd.make_context(
+                    with cmd.make_context(
                         name,
                         args,
                         parent=ctx,
                         allow_extra_args=True,
                         allow_interspersed_args=False,
                         resilient_parsing=True,
-                    )
-                    args = sub_ctx.args
+                    ) as sub_sub_ctx:
+                        args = sub_sub_ctx.args
+                        sub_ctx = sub_sub_ctx
 
                 ctx = sub_ctx
                 args = [*sub_ctx._protected_args, *sub_ctx.args]
