diff --git a/src/jinja2/filters.py b/src/jinja2/filters.py
index 80ea6504..5273606e 100644
--- a/src/jinja2/filters.py
+++ b/src/jinja2/filters.py
@@ -1163,6 +1163,7 @@ def sync_do_groupby(
     value: "t.Iterable[V]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     """Group a sequence of objects by an attribute using Python's
     :func:`itertools.groupby`. The attribute can use dot notation for
@@ -1203,6 +1204,21 @@ def sync_do_groupby(
           <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
         {% endfor %}</ul>
 
+    Like the :func:`~jinja-filters.sort` filter, sorting and grouping is
+    case-insensitive by default. The ``key`` for each group will use the
+    case of the first item encountered. For example, if a list of users
+    has cities ``["CA", "NY", "ca"]``, the "CA" and "ca" entries will be
+    grouped together.
+
+    .. sourcecode:: jinja
+
+        <ul>{% for city, items in users|groupby("city", case_sensitive=true) %}
+          <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
+        {% endfor %}</ul>
+
+    .. versionchanged:: 3.1
+        Added the ``case_sensitive`` parameter.
+
     .. versionchanged:: 3.0
         Added the ``default`` parameter.
 
@@ -1210,10 +1226,8 @@ def sync_do_groupby(
         The attribute supports dot notation for nested access.
     """
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, list(values))
-        for key, values in groupby(sorted(value, key=expr), expr)
-    ]
+    out = _group_by(value, expr, case_sensitive)
+    return out
 
 
 @async_variant(sync_do_groupby)  # type: ignore
@@ -1222,12 +1236,32 @@ async def do_groupby(
     value: "t.Union[t.AsyncIterable[V], t.Iterable[V]]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, await auto_to_list(values))
-        for key, values in groupby(sorted(await auto_to_list(value), key=expr), expr)
-    ]
+    out = _group_by(await auto_to_list(value), expr, case_sensitive)
+    return out
+
+
+def _group_by(
+    value: "t.Iterable[V]",
+    expr: "t.Callable[[t.Any], t.Any]",
+    case_sensitive: bool,
+) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
+    if case_sensitive:
+        key_func = expr
+    else:
+
+        def key_func(item: t.Any) -> t.Any:
+            return ignore_case(expr(item))
+
+    groups = []
+
+    for _key, values in groupby(sorted(value, key=key_func), key_func):
+        values = list(values)
+        groups.append(_GroupTuple(expr(values[0]), values))
+
+    return groups
 
 
 @pass_environment
