diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..4cf4b0e 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -67,6 +67,7 @@ __all__ = (
     'IsInstance',
     # other
     'FunctionCheck',
+    'IsHash',
     'IsJson',
     'IsUUID',
     # strings
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..14bca02 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -55,6 +55,55 @@ AnyJson = object
 JsonType = TypeVar('JsonType', AnyJson, Any)
 
 
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash of a given algorithm.
+    """
+
+    hash_lengths = {
+        'md5': 32,
+        'sha1': 40,
+        'sha224': 56,
+        'sha256': 64,
+        'sha384': 96,
+        'sha512': 128,
+    }
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm to use for the hash, must be one of `'md5'`, `'sha1'`, `'sha224'`, `'sha256'`,
+                `'sha384'`, or `'sha512'`.
+
+        ```py title="IsHash"
+        import hashlib
+        from dirty_equals import IsHash
+
+        assert hashlib.md5(b'foo').hexdigest() == IsHash('md5')
+        assert hashlib.sha1(b'foo').hexdigest() == IsHash('sha1')
+        assert hashlib.sha256(b'foo').hexdigest() == IsHash('sha256')
+        assert 'not_a_hash' != IsHash('md5')
+        ```
+        """
+        self.algorithm = algorithm
+        try:
+            self.length = self.hash_lengths[algorithm.lower()]
+        except KeyError:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        if len(other) != self.length:
+            return False
+        try:
+            int(other, 16)
+        except ValueError:
+            return False
+        return True
+
+
 class IsJson(DirtyEquals[JsonType]):
     """
     A class that checks if a value is a JSON object, and check the contents of the JSON.
