diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..291e95c 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsPhoneNumber, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -67,7 +67,9 @@ __all__ = (
     'IsInstance',
     # other
     'FunctionCheck',
+    'IsHash',
     'IsJson',
+    'IsPhoneNumber',
     'IsUUID',
     # strings
     'IsStr',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..2d89364 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -51,6 +51,56 @@ class IsUUID(DirtyEquals[UUID]):
             return True
 
 
+HashTypes = Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+HASH_LENGTHS: 'dict[HashTypes, int]' = {
+    'md5': 32,
+    'sha1': 40,
+    'sha224': 56,
+    'sha256': 64,
+    'sha384': 96,
+    'sha512': 128,
+}
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, for a given hashing algorithm.
+    """
+
+    def __init__(self, algorithm: HashTypes):
+        """
+        Args:
+            algorithm: The hashing algorithm used to generate the hash, one of `'md5'`, `'sha1'`, `'sha224'`,
+                `'sha256'`, `'sha384'` or `'sha512'`.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '5f4dcc3b5aa765d61d8327deb882cf99' == IsHash('md5')
+        assert '5baa61e4c9b93f3f0682250b6cf8331b7ee68fd8' == IsHash('sha1')
+        assert '5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8' == IsHash('sha256')
+        assert 'not-a-hash' != IsHash('sha256')
+        ```
+        """
+        algorithm_lower = algorithm.lower()
+        if algorithm_lower not in HASH_LENGTHS:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        self.algorithm: HashTypes = algorithm_lower  # type: ignore[assignment]
+        self.hash_length = HASH_LENGTHS[self.algorithm]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        if len(other) != self.hash_length:
+            return False
+        try:
+            int(other, 16)
+        except ValueError:
+            return False
+        return True
+
+
 AnyJson = object
 JsonType = TypeVar('JsonType', AnyJson, Any)
 
