diff --git a/src/jinja2/loaders.py b/src/jinja2/loaders.py
index d7d9bd04..20eabcf6 100644
--- a/src/jinja2/loaders.py
+++ b/src/jinja2/loaders.py
@@ -191,7 +191,7 @@ class FileSystemLoader(BaseLoader):
     def get_source(
         self, environment: "Environment", template: str
     ) -> t.Tuple[str, str, t.Callable[[], bool]]:
-        pieces = split_template_path(template)
+        pieces = split_template_path(os.path.normpath(template))
         for searchpath in self.searchpath:
             filename = os.path.join(searchpath, *pieces)
             f = open_if_exists(filename)
@@ -326,7 +326,8 @@ class PackageLoader(BaseLoader):
     def get_source(
         self, environment: "Environment", template: str
     ) -> t.Tuple[str, str, t.Optional[t.Callable[[], bool]]]:
-        p = os.path.join(self._template_root, *split_template_path(template))
+        pieces = split_template_path(os.path.normpath(template))
+        p = os.path.join(self._template_root, *pieces)
         up_to_date: t.Optional[t.Callable[[], bool]]
 
         if self._archive is None:
