diff --git a/outlines/backends/xgrammar.py b/outlines/backends/xgrammar.py
index fcb1672d..e7fb5c9f 100644
--- a/outlines/backends/xgrammar.py
+++ b/outlines/backends/xgrammar.py
@@ -1,6 +1,6 @@
 """Backend class for XGrammar."""
 
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Optional
 
 from outlines.backends.base import BaseBackend
 from outlines.models import SteerableModel
@@ -9,6 +9,7 @@ from outlines.processors.base_logits_processor import (
     OutlinesLogitsProcessor,
     TensorType
 )
+from outlines.processors.tensor_adapters import TensorAdapter
 
 if TYPE_CHECKING:
     from xgrammar.contrib.hf import LogitsProcessor
@@ -22,12 +23,26 @@ class XGrammarLogitsProcessor(OutlinesLogitsProcessor):
 
     """
 
-    def __init__(self, compiled_grammar: str):
+    def __init__(
+        self,
+        compiled_grammar: str,
+        tensor_library_name: str = "torch",
+        custom_adapter: Optional[TensorAdapter] = None,
+    ):
         """
         Parameters
         ----------
         compiled_grammar: str
             The compiled grammar to use to create the logits processor.
+        tensor_library_name: str
+            The name of the tensor library to use. Possible values are
+            "torch", "mlx", "numpy", "jax", "tensorflow" and "custom". Use
+            "custom" along with the `custom_adapter` argument to provide
+            your own tensor adapter implementation.
+        custom_adapter: Optional[TensorAdapter]
+            An instance of a `TensorAdapter` subclass to use to manipulate
+            tensors. Must be provided if and only if `tensor_library_name`
+            is "custom".
 
         """
         import xgrammar as xgr
@@ -35,7 +50,7 @@ class XGrammarLogitsProcessor(OutlinesLogitsProcessor):
         self.xgr = xgr
         self.compiled_grammar = compiled_grammar
         self.xgrammar_logits_processor = None
-        super().__init__("torch")
+        super().__init__(tensor_library_name, custom_adapter)
 
     def reset(self):
         """Reset the logits processor for a new generation."""
@@ -79,7 +94,10 @@ class XGrammarBackend(BaseBackend):
         self.grammar_compiler = xgr.GrammarCompiler(tokenizer_info)
 
     def get_json_schema_logits_processor(
-        self, json_schema: str
+        self,
+        json_schema: str,
+        tensor_library_name: str = "torch",
+        custom_adapter: Optional[TensorAdapter] = None,
     ) -> "LogitsProcessor":
         """Create a logits processor from a JSON schema.
 
@@ -87,6 +105,13 @@ class XGrammarBackend(BaseBackend):
         ----------
         json_schema: str
             The JSON schema to create a logits processor from.
+        tensor_library_name: str
+            The name of the tensor library to use. Possible values are
+            "torch", "mlx", "numpy", "jax", "tensorflow" and "custom".
+        custom_adapter: Optional[TensorAdapter]
+            An instance of a `TensorAdapter` subclass to use to manipulate
+            tensors. Must be provided if and only if `tensor_library_name`
+            is "custom".
 
         Returns
         -------
@@ -97,15 +122,29 @@ class XGrammarBackend(BaseBackend):
         compiled_grammar = self.grammar_compiler.compile_json_schema(
             json_schema
         )
-        return XGrammarLogitsProcessor(compiled_grammar)
+        return XGrammarLogitsProcessor(
+            compiled_grammar, tensor_library_name, custom_adapter
+        )
 
-    def get_regex_logits_processor(self, regex: str) -> "LogitsProcessor":
+    def get_regex_logits_processor(
+        self,
+        regex: str,
+        tensor_library_name: str = "torch",
+        custom_adapter: Optional[TensorAdapter] = None,
+    ) -> "LogitsProcessor":
         """Create a logits processor from a regex.
 
         Parameters
         ----------
         regex: str
             The regex to create a logits processor from.
+        tensor_library_name: str
+            The name of the tensor library to use. Possible values are
+            "torch", "mlx", "numpy", "jax", "tensorflow" and "custom".
+        custom_adapter: Optional[TensorAdapter]
+            An instance of a `TensorAdapter` subclass to use to manipulate
+            tensors. Must be provided if and only if `tensor_library_name`
+            is "custom".
 
         Returns
         -------
@@ -114,15 +153,29 @@ class XGrammarBackend(BaseBackend):
 
         """
         compiled_grammar = self.grammar_compiler.compile_regex(regex)
-        return XGrammarLogitsProcessor(compiled_grammar)
+        return XGrammarLogitsProcessor(
+            compiled_grammar, tensor_library_name, custom_adapter
+        )
 
-    def get_cfg_logits_processor(self, grammar: str) -> "LogitsProcessor":
+    def get_cfg_logits_processor(
+        self,
+        grammar: str,
+        tensor_library_name: str = "torch",
+        custom_adapter: Optional[TensorAdapter] = None,
+    ) -> "LogitsProcessor":
         """Create a logits processor from a context-free grammar.
 
         Parameters
         ----------
         grammar: str
             The context-free grammar to create a logits processor from.
+        tensor_library_name: str
+            The name of the tensor library to use. Possible values are
+            "torch", "mlx", "numpy", "jax", "tensorflow" and "custom".
+        custom_adapter: Optional[TensorAdapter]
+            An instance of a `TensorAdapter` subclass to use to manipulate
+            tensors. Must be provided if and only if `tensor_library_name`
+            is "custom".
 
         Returns
         -------
@@ -131,4 +184,6 @@ class XGrammarBackend(BaseBackend):
 
         """
         compiled_grammar = self.grammar_compiler.compile_grammar(grammar)
-        return XGrammarLogitsProcessor(compiled_grammar)
+        return XGrammarLogitsProcessor(
+            compiled_grammar, tensor_library_name, custom_adapter
+        )
diff --git a/outlines/processors/base_logits_processor.py b/outlines/processors/base_logits_processor.py
index 9b4e4fe5..90f06618 100644
--- a/outlines/processors/base_logits_processor.py
+++ b/outlines/processors/base_logits_processor.py
@@ -1,9 +1,10 @@
 """Base class for logits processors."""
 
 from abc import abstractmethod
-from typing import TypeVar
+from typing import Optional, TypeVar
 
 from outlines.processors.tensor_adapters import (
+    TensorAdapter,
     TensorAdapterImplementation,
     tensor_adapters,
 )
@@ -22,15 +23,39 @@ class OutlinesLogitsProcessor:
     """
     tensor_adapter: TensorAdapterImplementation
 
-    def __init__(self, tensor_library_name: str):
+    def __init__(
+        self,
+        tensor_library_name: str,
+        custom_adapter: Optional[TensorAdapter] = None,
+    ):
         """
         Parameters
         ----------
         tensor_library_name
             The name of the library to use to manipulate tensors. Possible
-            values are "jax", "mlx", "numpy", "tensorflow" and "torch". You
-            must choose the library that your model is using.
+            values are "jax", "mlx", "numpy", "tensorflow", "torch" and
+            "custom". You must choose the library that your model is using.
+            Use "custom" along with the `custom_adapter` argument to provide
+            your own tensor adapter implementation.
+        custom_adapter
+            An instance of a `TensorAdapter` subclass to use to manipulate
+            tensors. Must be provided if and only if `tensor_library_name` is
+            "custom".
         """
+        if tensor_library_name == "custom":
+            if custom_adapter is None:
+                raise ValueError(
+                    "custom_adapter must be provided when tensor_library_name "
+                    "is 'custom'"
+                )
+            self.tensor_adapter = custom_adapter
+            return
+        elif custom_adapter is not None:
+            raise ValueError(
+                "custom_adapter should only be provided when "
+                "tensor_library_name is 'custom'"
+            )
+
         # Temporary fix as torch raises a warning that can cause can an error
         # with python 3.12.
         if tensor_library_name == "torch":
diff --git a/outlines/processors/tensor_adapters/__init__.py b/outlines/processors/tensor_adapters/__init__.py
index 8e3afe7f..474fe9e3 100644
--- a/outlines/processors/tensor_adapters/__init__.py
+++ b/outlines/processors/tensor_adapters/__init__.py
@@ -2,6 +2,7 @@
 
 from typing import Union
 
+from .base import TensorAdapter
 from .jax import JAXTensorAdapter
 from .mlx import MLXTensorAdapter
 from .numpy import NumpyTensorAdapter
