diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..4cf4b0e 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -67,6 +67,7 @@ __all__ = (
     'IsInstance',
     # other
     'FunctionCheck',
+    'IsHash',
     'IsJson',
     'IsUUID',
     # strings
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..89824ab 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,5 +1,6 @@
 import json
-from typing import Any, Callable, TypeVar, overload
+import re
+from typing import Any, Callable, Dict, TypeVar, overload
 from uuid import UUID
 
 from ._base import DirtyEquals
@@ -51,6 +52,50 @@ class IsUUID(DirtyEquals[UUID]):
             return True
 
 
+HashTypes = Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+
+hash_lengths: Dict[str, int] = {
+    'md5': 32,
+    'sha1': 40,
+    'sha224': 56,
+    'sha256': 64,
+    'sha384': 96,
+    'sha512': 128,
+}
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, does not check the value is correct.
+    """
+
+    def __init__(self, algorithm: HashTypes):
+        """
+        Args:
+            algorithm: The algorithm to use to check the hash, one of `'md5'`, `'sha1'`, `'sha224'`, `'sha256'`,
+                `'sha384'` or `'sha512'`.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '5eb63bbbe01eeed093cb22bb8f5acdc3' == IsHash('md5')
+        assert 'd2d2d2df7c3bcc9021bf6a3818bb5cbfe0058e37' == IsHash('sha1')
+        assert '3f786850e387550fdab836ed7e6dc881de23001b' != IsHash('md5')
+        ```
+        """
+        self.algorithm = algorithm
+        try:
+            self.length = hash_lengths[algorithm.lower()]
+        except KeyError:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        return bool(re.fullmatch(f'[0-9a-f]{{{self.length}}}', other, flags=re.IGNORECASE))
+
+
 AnyJson = object
 JsonType = TypeVar('JsonType', AnyJson, Any)
 
