diff --git a/Tests/test_image.py b/Tests/test_image.py
index c2e850c36..409c32b2b 100644
--- a/Tests/test_image.py
+++ b/Tests/test_image.py
@@ -182,6 +182,20 @@ class TestImage:
         with Image.open(out) as reloaded:
             assert_image_equal(im, reloaded)
 
+    def test_dry_run(self, tmp_path: Path) -> None:
+        out = tmp_path / "temp.png"
+        im = hopper()
+        im.save(out, dry_run=True)
+        assert not out.exists()
+
+        im.save(out)
+        assert out.exists()
+
+    def test_dry_run_raises_error(self) -> None:
+        im = hopper("RGBA")
+        with pytest.raises(OSError):
+            im.save(io.BytesIO(), format="JPEG", dry_run=True)
+
     def test_pathlib(self, tmp_path: Path) -> None:
         with Image.open(Path("Tests/images/multipage-mmap.tiff")) as im:
             assert im.mode == "P"
diff --git a/src/PIL/Image.py b/src/PIL/Image.py
index 233df592c..9df10dc6d 100644
--- a/src/PIL/Image.py
+++ b/src/PIL/Image.py
@@ -2505,6 +2505,10 @@ class Image:
              red = Image.new("RGB", (1, 1), "#f00")
              red.encoderinfo = {"xmp": b"test"}
              black.save("out.mpo", save_all=True, append_images=[red])
+        :param dry_run: If true, the image is processed for saving, including
+           any format conversion and encoding, but the result is not written
+           to disk. This can be used to validate that an image can be saved
+           without the overhead of disk I/O.
         :returns: None
         :exception ValueError: If the output format could not be determined
            from the file name.  Use the format option to solve this.
@@ -2543,6 +2547,7 @@ class Image:
         # may mutate self!
         self._ensure_mutable()
 
+        dry_run = params.pop("dry_run", False)
         save_all = params.pop("save_all", None)
         self.encoderinfo = {**getattr(self, "encoderinfo", {}), **params}
         self.encoderconfig: tuple[Any, ...] = ()
@@ -2559,7 +2564,9 @@ class Image:
             save_handler = SAVE[format.upper()]
 
         created = False
-        if open_fp:
+        if dry_run:
+            fp = io.BytesIO()
+        elif open_fp:
             created = not os.path.exists(filename)
             if params.get("append", False):
                 # Open also for reading ("+"), because TIFF save_all
