diff --git a/crates/typst/src/foundations/str.rs b/crates/typst/src/foundations/str.rs
index d90b6f206..bc76a3d58 100644
--- a/crates/typst/src/foundations/str.rs
+++ b/crates/typst/src/foundations/str.rs
@@ -177,25 +177,44 @@ impl Str {
     }
 
     /// Extracts the first grapheme cluster of the string.
-    /// Fails with an error if the string is empty.
+    /// Fails with an error if the string is empty or has fewer grapheme
+    /// clusters than `repeat`.
     #[func]
-    pub fn first(&self) -> StrResult<Str> {
-        self.0
-            .graphemes(true)
-            .next()
-            .map(Into::into)
-            .ok_or_else(string_is_empty)
+    pub fn first(
+        &self,
+        /// Number of times to repeat this operation. Fails if exceeds string length.
+        #[named]
+        #[default(1)]
+        repeat: usize,
+    ) -> StrResult<Str> {
+        let mut end = 0;
+        let mut iter = self.0.graphemes(true);
+        for _ in 0..repeat {
+            end += iter.next().ok_or_else(|| string_too_short(self.0.is_empty()))?.len();
+        }
+        Ok(self.0[..end].into())
     }
 
     /// Extracts the last grapheme cluster of the string.
-    /// Fails with an error if the string is empty.
+    /// Fails with an error if the string is empty or has fewer grapheme
+    /// clusters than `repeat`.
     #[func]
-    pub fn last(&self) -> StrResult<Str> {
-        self.0
-            .graphemes(true)
-            .next_back()
-            .map(Into::into)
-            .ok_or_else(string_is_empty)
+    pub fn last(
+        &self,
+        /// Number of times to repeat this operation. Fails if exceeds string length.
+        #[named]
+        #[default(1)]
+        repeat: usize,
+    ) -> StrResult<Str> {
+        let mut start = self.0.len();
+        let mut iter = self.0.graphemes(true);
+        for _ in 0..repeat {
+            start -= iter
+                .next_back()
+                .ok_or_else(|| string_too_short(self.0.is_empty()))?
+                .len();
+        }
+        Ok(self.0[start..].into())
     }
 
     /// Extracts the first grapheme cluster after the specified index. Returns
@@ -835,6 +854,17 @@ fn string_is_empty() -> EcoString {
     "string is empty".into()
 }
 
+/// The error message when the string has fewer grapheme clusters than
+/// requested by `repeat`.
+#[cold]
+fn string_too_short(is_empty: bool) -> EcoString {
+    if is_empty {
+        string_is_empty()
+    } else {
+        "string is not long enough for the given `repeat`".into()
+    }
+}
+
 /// A regular expression.
 ///
 /// Can be used as a [show rule selector]($styling/#show-rules) and with
