diff --git a/llama-index-core/llama_index/core/base/llms/types.py b/llama-index-core/llama_index/core/base/llms/types.py
index 00e30cf23..b079aeb2f 100644
--- a/llama-index-core/llama_index/core/base/llms/types.py
+++ b/llama-index-core/llama_index/core/base/llms/types.py
@@ -75,14 +75,29 @@ class ImageBlock(BaseModel):
         if not self.image:
             return self
 
-        if not self.image_mimetype:
-            guess = filetype.guess(self.image)
-            self.image_mimetype = guess.mime if guess else None
+        self.get_image_mimetype()
 
         self.image = base64.b64encode(self.image)
 
         return self
 
+    def get_image_mimetype(self) -> str | None:
+        """Return the image mimetype, guessing and caching it if not already set.
+
+        To avoid resource-intense operations, this won't load the path or the
+        URL to guess the mimetype -- it only inspects `self.image` if present.
+        """
+        if self.image_mimetype:
+            return self.image_mimetype
+
+        if not self.image:
+            return None
+
+        guess = filetype.guess(self.image)
+        self.image_mimetype = guess.mime if guess else None
+
+        return self.image_mimetype
+
     def resolve_image(self, as_base64: bool = False) -> BytesIO:
         """Resolve an image such that PIL can read it.
 
