diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..717e916 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,5 +1,6 @@
 import json
-from typing import Any, Callable, TypeVar, overload
+import re
+from typing import Any, Callable, Dict, TypeVar, overload
 from uuid import UUID
 
 from ._base import DirtyEquals
@@ -120,6 +121,49 @@ class IsJson(DirtyEquals[JsonType]):
             return False
 
 
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, specifically checking the length and characters used.
+    """
+
+    exact_length: Dict[str, int] = {
+        'md5': 32,
+        'sha1': 40,
+        'sha224': 56,
+        'sha256': 64,
+        'sha384': 96,
+        'sha512': 128,
+    }
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm to use for the hash, must be one of the supported algorithms.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '098f6bcd4621d373cade4e832627b4f6' == IsHash('md5')
+        assert 'not-a-hash' != IsHash('md5')
+        assert 'a94a8fe5ccb19ba61c4c0873d391e987982fbbd3' == IsHash('sha1')
+        assert '9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08' == IsHash('sha256')
+        ```
+        """
+        algorithm_key = algorithm.lower()
+        if algorithm_key not in self.exact_length:
+            raise ValueError(f'Unknown hash algorithm {algorithm!r}')
+        self.algorithm = algorithm
+        self.hash_length = self.exact_length[algorithm_key]
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        if len(other) != self.hash_length:
+            return False
+        return bool(re.fullmatch('[0-9a-fA-F]+', other))
+
+
 class FunctionCheck(DirtyEquals[Any]):
     """
     Use a function to check if a value "equals" whatever you want to check
