diff --git a/outlines/types/__init__.py b/outlines/types/__init__.py
index 85af9adc..d12be3c2 100644
--- a/outlines/types/__init__.py
+++ b/outlines/types/__init__.py
@@ -48,6 +48,9 @@ char = Regex(r"\w")
 newline = Regex(r"(\r\n|\r|\n)")  # Matched new lines on Linux, Windows & MacOS
 whitespace = Regex(r"\s")
 
+# Cryptographic hash types
+hash_sha256 = Regex(r"[a-fA-F0-9]{64}")
+
 # Document-specific types
 sentence = Regex(r"[A-Z].*\s*[.!?]")
 paragraph = Regex(rf"{sentence.pattern}(?:\s+{sentence.pattern})*\n+")
