diff --git a/outlines/types/__init__.py b/outlines/types/__init__.py
index 85af9adc..a7ebd55e 100644
--- a/outlines/types/__init__.py
+++ b/outlines/types/__init__.py
@@ -52,6 +52,9 @@ whitespace = Regex(r"\s")
 sentence = Regex(r"[A-Z].*\s*[.!?]")
 paragraph = Regex(rf"{sentence.pattern}(?:\s+{sentence.pattern})*\n+")
 
+# Cryptographic hash types
+hash_sha256 = Regex(r"[a-fA-F0-9]{64}")
+
 
 # The following regex is FRC 5322 compliant and was found at:
 # https://emailregex.com/
