diff --git a/Tests/test_image.py b/Tests/test_image.py
index c2e850c36..0b5fdf09f 100644
--- a/Tests/test_image.py
+++ b/Tests/test_image.py
@@ -258,6 +258,32 @@ class TestImage:
             assert im.readonly
             im.save(temp_file)
 
+    def test_preserve_readonly(self, tmp_path: Path) -> None:
+        temp_file = tmp_path / "temp.bmp"
+
+        data = bytearray(b"\xff")
+        im = Image.frombuffer("L", (1, 1), data, "raw", "L", 0, 1)
+        assert im.readonly
+
+        im.save(temp_file, preserve_readonly=True)
+        assert im.readonly
+
+        data[0] = 64
+        assert im.getpixel((0, 0)) == 64
+
+    def test_preserve_readonly_false(self, tmp_path: Path) -> None:
+        temp_file = tmp_path / "temp.bmp"
+
+        data = bytearray(b"\xff")
+        im = Image.frombuffer("L", (1, 1), data, "raw", "L", 0, 1)
+        assert im.readonly
+
+        im.save(temp_file)
+        assert not im.readonly
+
+        data[0] = 64
+        assert im.getpixel((0, 0)) == 255
+
     def test_dump(self, tmp_path: Path) -> None:
         im = Image.new("L", (10, 10))
         im._dump(str(tmp_path / "temp_L.ppm"))
diff --git a/src/PIL/Image.py b/src/PIL/Image.py
index 233df592c..7a38cd4ea 100644
--- a/src/PIL/Image.py
+++ b/src/PIL/Image.py
@@ -2467,7 +2467,12 @@ class Image:
         )
 
     def save(
-        self, fp: StrOrBytesPath | IO[bytes], format: str | None = None, **params: Any
+        self,
+        fp: StrOrBytesPath | IO[bytes],
+        format: str | None = None,
+        *,
+        preserve_readonly: bool = False,
+        **params: Any,
     ) -> None:
         """
         Saves this image under the given filename.  If no format is
@@ -2490,6 +2495,11 @@ class Image:
            format to use is determined from the filename extension.
            If a file object was used instead of a filename, this
            parameter should always be used.
+        :param preserve_readonly: If true, the image's readonly state, e.g. as
+           set by :py:meth:`~PIL.Image.frombuffer`, will be preserved,
+           keeping it linked to the underlying data. If false, the default,
+           the image will be made mutable before saving, which may involve
+           copying the underlying data.
         :param params: Extra parameters to the image writer. These can also be
            set on the image itself through ``encoderinfo``. This is useful when
            saving multiple images::
@@ -2541,7 +2551,10 @@ class Image:
                 raise ValueError(msg) from e
 
         # may mutate self!
-        self._ensure_mutable()
+        if preserve_readonly:
+            self.load()
+        else:
+            self._ensure_mutable()
 
         save_all = params.pop("save_all", None)
         self.encoderinfo = {**getattr(self, "encoderinfo", {}), **params}
