diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..2757f10 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,4 +1,5 @@
 import json
+import re
 from typing import Any, Callable, TypeVar, overload
 from uuid import UUID
 
@@ -120,6 +121,51 @@ class IsJson(DirtyEquals[JsonType]):
             return False
 
 
+HashAlgorithm = Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+
+_hash_lengths = {
+    'md5': 32,
+    'sha1': 40,
+    'sha224': 56,
+    'sha256': 64,
+    'sha384': 96,
+    'sha512': 128,
+}
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash for a specific hashing algorithm, using regular expressions.
+    """
+
+    def __init__(self, algorithm: HashAlgorithm):
+        """
+        Args:
+            algorithm: The algorithm to validate against, one of 'md5', 'sha1', 'sha224', 'sha256', 'sha384', or
+                'sha512'.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert '5eb63bbbe01eeed093cb22bb8f5acdc3' == IsHash('md5')
+        assert 'd0be2dc421be4fcd0172e5afceea3970e2f3d940' == IsHash('sha1')
+        assert '2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae' != IsHash('sha256')
+        ```
+        """
+        self.algorithm = algorithm
+        self._lower_algorithm = algorithm.lower()
+        if self._lower_algorithm not in _hash_lengths:
+            raise ValueError(f'Unknown algorithm {algorithm!r}')
+        self._expected_length = _hash_lengths[self._lower_algorithm]
+        self._regex = re.compile(f'^[0-9a-fA-F]{{{self._expected_length}}}$')
+        super().__init__(algorithm)
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        return bool(self._regex.fullmatch(other))
+
+
 class FunctionCheck(DirtyEquals[Any]):
     """
     Use a function to check if a value "equals" whatever you want to check
