diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..c667730 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -120,6 +120,59 @@ class IsJson(DirtyEquals[JsonType]):
             return False
 
 
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid hash, for a given hashing algorithm.
+    """
+
+    def __init__(self, algorithm: Literal['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']):
+        """
+        Args:
+            algorithm: The algorithm to use for the hash, must be one of
+                `'md5'`, `'sha1'`, `'sha224'`, `'sha256'`, `'sha384'` or `'sha512'`.
+
+        ```py title="IsHash"
+        import hashlib
+        from dirty_equals import IsHash
+
+        assert hashlib.md5(b'dirty-equals').hexdigest() == IsHash('md5')
+        assert hashlib.sha1(b'dirty-equals').hexdigest() == IsHash('sha1')
+        assert hashlib.sha256(b'dirty-equals').hexdigest() == IsHash('sha256')
+        assert 'not-a-hash' != IsHash('sha256')
+        ```
+        """
+        self.algorithm = algorithm.lower()
+        self.hash_length = self._get_hash_algorithm_length()
+        super().__init__(algorithm)
+
+    def _get_hash_algorithm_length(self) -> int:
+        if self.algorithm == 'md5':
+            return 32
+        elif self.algorithm == 'sha1':
+            return 40
+        elif self.algorithm == 'sha224':
+            return 56
+        elif self.algorithm == 'sha256':
+            return 64
+        elif self.algorithm == 'sha384':
+            return 96
+        elif self.algorithm == 'sha512':
+            return 128
+        else:
+            raise ValueError(f'Unknown hash algorithm {self.algorithm!r}')
+
+    def equals(self, other: Any) -> bool:
+        if not isinstance(other, str):
+            return False
+        if len(other) != self.hash_length:
+            return False
+        try:
+            int(other, 16)
+        except ValueError:
+            return False
+        return True
+
+
 class FunctionCheck(DirtyEquals[Any]):
     """
     Use a function to check if a value "equals" whatever you want to check
