diff --git a/src/jinja2/filters.py b/src/jinja2/filters.py
index 80ea6504..9731074d 100644
--- a/src/jinja2/filters.py
+++ b/src/jinja2/filters.py
@@ -1163,6 +1163,7 @@ def sync_do_groupby(
     value: "t.Iterable[V]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     """Group a sequence of objects by an attribute using Python's
     :func:`itertools.groupby`. The attribute can use dot notation for
@@ -1203,18 +1204,35 @@ def sync_do_groupby(
           <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
         {% endfor %}</ul>
 
+    You can specify ``case_sensitive=True`` to disable case-insensitive
+    grouping, which is the default.
+
+    .. versionchanged:: 3.1
+        Added the ``case_sensitive`` parameter.
+
     .. versionchanged:: 3.0
         Added the ``default`` parameter.
 
     .. versionchanged:: 2.6
         The attribute supports dot notation for nested access.
     """
-    expr = make_attrgetter(environment, attribute, default=default)
-    return [
+    expr = make_attrgetter(
+        environment,
+        attribute,
+        postprocess=ignore_case if not case_sensitive else None,
+        default=default,
+    )
+    out = [
         _GroupTuple(key, list(values))
         for key, values in groupby(sorted(value, key=expr), expr)
     ]
 
+    if not case_sensitive:
+        raw_expr = make_attrgetter(environment, attribute, default=default)
+        out = [_GroupTuple(raw_expr(values[0]), values) for _, values in out]
+
+    return out
+
 
 @async_variant(sync_do_groupby)  # type: ignore
 async def do_groupby(
@@ -1222,13 +1240,25 @@ async def do_groupby(
     value: "t.Union[t.AsyncIterable[V], t.Iterable[V]]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
-    expr = make_attrgetter(environment, attribute, default=default)
-    return [
+    expr = make_attrgetter(
+        environment,
+        attribute,
+        postprocess=ignore_case if not case_sensitive else None,
+        default=default,
+    )
+    out = [
         _GroupTuple(key, await auto_to_list(values))
         for key, values in groupby(sorted(await auto_to_list(value), key=expr), expr)
     ]
 
+    if not case_sensitive:
+        raw_expr = make_attrgetter(environment, attribute, default=default)
+        out = [_GroupTuple(raw_expr(values[0]), values) for _, values in out]
+
+    return out
+
 
 @pass_environment
 def sync_do_sum(
