diff --git a/dirty_equals/__init__.py b/dirty_equals/__init__.py
index ba8214e..86b0ee9 100644
--- a/dirty_equals/__init__.py
+++ b/dirty_equals/__init__.py
@@ -18,7 +18,7 @@ from ._numeric import (
     IsPositiveFloat,
     IsPositiveInt,
 )
-from ._other import FunctionCheck, IsJson, IsUUID
+from ._other import FunctionCheck, IsHash, IsJson, IsUUID
 from ._sequence import Contains, HasLen, IsList, IsListOrTuple, IsTuple
 from ._strings import IsAnyStr, IsBytes, IsStr
 
@@ -69,6 +69,7 @@ __all__ = (
     'FunctionCheck',
     'IsJson',
     'IsUUID',
+    'IsHash',
     # strings
     'IsStr',
     'IsBytes',
diff --git a/dirty_equals/_other.py b/dirty_equals/_other.py
index b2b8379..ed9f62a 100644
--- a/dirty_equals/_other.py
+++ b/dirty_equals/_other.py
@@ -1,4 +1,5 @@
 import json
+import re
 from typing import Any, Callable, TypeVar, overload
 from uuid import UUID
 
@@ -120,6 +121,53 @@ class IsJson(DirtyEquals[JsonType]):
             return False
 
 
+HashTypes = Literal['md5', 'sha-1', 'sha-256']
+
+
+class IsHash(DirtyEquals[str]):
+    """
+    A class that checks if a value is a valid common hash type, using a simple length and allowed characters regex.
+    """
+
+    def __init__(self, hash_type: HashTypes):
+        """
+        Args:
+            hash_type: The hash type to check. Must be specified.
+
+        ```py title="IsHash"
+        from dirty_equals import IsHash
+
+        assert 'f1e069787ece74531d112559945c6871' == IsHash('md5')
+        assert b'f1e069787ece74531d112559945c6871' == IsHash('md5')
+        assert 'f1e069787ece74531d112559945c6871' != IsHash('sha-256')
+        assert 'F1E069787ECE74531D112559945C6871' == IsHash('md5')
+        assert '40bd001563085fc35165329ea1ff5c5ecbdbbeef' == IsHash('sha-1')
+        assert 'a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3' == IsHash('sha-256')
+        ```
+        """
+
+        allowed_hashes = HashTypes.__args__  # type: ignore[attr-defined]
+        if hash_type not in allowed_hashes:
+            raise ValueError(f"Hash type must be one of the following values: {', '.join(allowed_hashes)}")
+
+        self.hash_type = hash_type
+        super().__init__(hash_type)
+
+    def equals(self, other: Any) -> bool:
+        if isinstance(other, str):
+            s = other
+        elif isinstance(other, (bytes, bytearray)):
+            s = other.decode()
+        else:
+            return False
+        hash_type_regex_patterns = {
+            'md5': r'[a-fA-F\d]{32}',
+            'sha-1': r'[a-fA-F\d]{40}',
+            'sha-256': r'[a-fA-F\d]{64}',
+        }
+        return bool(re.fullmatch(hash_type_regex_patterns[self.hash_type], s))
+
+
 class FunctionCheck(DirtyEquals[Any]):
     """
     Use a function to check if a value "equals" whatever you want to check
