diff --git a/crates/typst/src/foundations/str.rs b/crates/typst/src/foundations/str.rs
index d90b6f206..7c6116648 100644
--- a/crates/typst/src/foundations/str.rs
+++ b/crates/typst/src/foundations/str.rs
@@ -179,23 +179,71 @@ impl Str {
     /// Extracts the first grapheme cluster of the string.
     /// Fails with an error if the string is empty.
     #[func]
-    pub fn first(&self) -> StrResult<Str> {
-        self.0
-            .graphemes(true)
-            .next()
-            .map(Into::into)
-            .ok_or_else(string_is_empty)
+    pub fn first(
+        &self,
+        /// Number of times to repeat this operation. Fails if exceeds string length.
+        #[named]
+        #[default(1)]
+        repeat: usize,
+        /// A set of characters to strip from the start of the string before
+        /// extracting the first grapheme cluster(s).
+        #[named]
+        strip: Option<Str>,
+    ) -> StrResult<Str> {
+        let s = match &strip {
+            Some(strip) => self.0.trim_start_matches(|c| strip.as_str().contains(c)),
+            None => self.0.as_str(),
+        };
+        let mut graphemes = s.graphemes(true);
+        let mut count = 0;
+        for _ in 0..repeat {
+            if graphemes.next().is_none() {
+                if count == 0 {
+                    return Err(string_is_empty());
+                }
+                return Err(eco_format!(
+                    "the string only has {count} grapheme cluster(s), not {repeat}"
+                ));
+            }
+            count += 1;
+        }
+        let end = s.len() - graphemes.as_str().len();
+        Ok(s[..end].into())
     }
 
     /// Extracts the last grapheme cluster of the string.
     /// Fails with an error if the string is empty.
     #[func]
-    pub fn last(&self) -> StrResult<Str> {
-        self.0
-            .graphemes(true)
-            .next_back()
-            .map(Into::into)
-            .ok_or_else(string_is_empty)
+    pub fn last(
+        &self,
+        /// Number of times to repeat this operation. Fails if exceeds string length.
+        #[named]
+        #[default(1)]
+        repeat: usize,
+        /// A set of characters to strip from the end of the string before
+        /// extracting the last grapheme cluster(s).
+        #[named]
+        strip: Option<Str>,
+    ) -> StrResult<Str> {
+        let s = match &strip {
+            Some(strip) => self.0.trim_end_matches(|c| strip.as_str().contains(c)),
+            None => self.0.as_str(),
+        };
+        let mut graphemes = s.graphemes(true);
+        let mut count = 0;
+        for _ in 0..repeat {
+            if graphemes.next_back().is_none() {
+                if count == 0 {
+                    return Err(string_is_empty());
+                }
+                return Err(eco_format!(
+                    "the string only has {count} grapheme cluster(s), not {repeat}"
+                ));
+            }
+            count += 1;
+        }
+        let start = graphemes.as_str().len();
+        Ok(s[start..].into())
     }
 
     /// Extracts the first grapheme cluster after the specified index. Returns
