diff --git a/src/jinja2/filters.py b/src/jinja2/filters.py
index 80ea6504..45f8ec31 100644
--- a/src/jinja2/filters.py
+++ b/src/jinja2/filters.py
@@ -1163,6 +1163,7 @@ def sync_do_groupby(
     value: "t.Iterable[V]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     """Group a sequence of objects by an attribute using Python's
     :func:`itertools.groupby`. The attribute can use dot notation for
@@ -1203,6 +1204,17 @@ def sync_do_groupby(
           <li>{{ city }}: {{ items|map(attribute="name")|join(", ") }}</li>
         {% endfor %}</ul>
 
+    Like the :func:`~jinja-filters.sort` filter, sorting and grouping is
+    case-insensitive by default. The ``key`` for each group will use the
+    case of the first item encountered for that group. For example, if a
+    dataset contains ``"CA"`` and ``"ca"``, the final result will use the
+    grouping of whichever version was encountered first. Set
+    ``case_sensitive`` to `True` to group multiple differently cased
+    versions of the same string distinctly.
+
+    .. versionchanged:: 3.1
+        Added the ``case_sensitive`` parameter.
+
     .. versionchanged:: 3.0
         Added the ``default`` parameter.
 
@@ -1210,10 +1222,22 @@ def sync_do_groupby(
         The attribute supports dot notation for nested access.
     """
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, list(values))
-        for key, values in groupby(sorted(value, key=expr), expr)
-    ]
+
+    def key_func(item: t.Any) -> t.Any:
+        key = expr(item)
+
+        if not case_sensitive:
+            key = ignore_case(key)
+
+        return key
+
+    out = []
+
+    for _, values in groupby(sorted(value, key=key_func), key_func):
+        values = list(values)
+        out.append(_GroupTuple(expr(values[0]), values))
+
+    return out
 
 
 @async_variant(sync_do_groupby)  # type: ignore
@@ -1222,12 +1246,25 @@ async def do_groupby(
     value: "t.Union[t.AsyncIterable[V], t.Iterable[V]]",
     attribute: t.Union[str, int],
     default: t.Optional[t.Any] = None,
+    case_sensitive: bool = False,
 ) -> "t.List[t.Tuple[t.Any, t.List[V]]]":
     expr = make_attrgetter(environment, attribute, default=default)
-    return [
-        _GroupTuple(key, await auto_to_list(values))
-        for key, values in groupby(sorted(await auto_to_list(value), key=expr), expr)
-    ]
+
+    def key_func(item: t.Any) -> t.Any:
+        key = expr(item)
+
+        if not case_sensitive:
+            key = ignore_case(key)
+
+        return key
+
+    out = []
+
+    for _, values in groupby(sorted(await auto_to_list(value), key=key_func), key_func):
+        values = await auto_to_list(values)
+        out.append(_GroupTuple(expr(values[0]), values))
+
+    return out
 
 
 @pass_environment
